{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "########################################################################\n",
    "# File: Chan_Nicholas_missingMotif.ipynb\n",
    "# Purpose: to find hidden motifs\n",
    "#   main(infile='FILE_PATH',outfile='FILE_PATH', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '0'])\n",
    "#\n",
    "# Author: Nicholas Chan\n",
    "# History: 10/18/2021 Created\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command Line Class\n",
    "Provided by Dr. B for parsing command line arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "\n",
    "\n",
    "class CommandLine():\n",
    "    \"\"\"\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse,\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination exception Usage.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, inOpts = None):\n",
    "        \"\"\"\n",
    "        CommandLine constructor.\n",
    "        \n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        \"\"\"\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "            )\n",
    "\n",
    "        self.parser.add_argument('-m', '--minMotif', type=int, default=3, action='store', help='minimum motif length to consider for Markov Model')\n",
    "        self.parser.add_argument('-M', '--maxMotif', type=int, default=8, action='store', help='maximum motif length to consider for Markov Model')\n",
    "        self.parser.add_argument('-c', '--cutoff', type=int, default=-4, action='store', help='Significant Z-score cutoff value')\n",
    "        # Command line option to use Gibbs sampling to find the optimal consensus motif.\n",
    "#         self.parser.add_argument('-g', '--gibbsampling', type=float, default=1, action='store', help='implement Gibbs sampling')\n",
    "\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FastAreader Class\n",
    "Provided by Dr. B for reading Fasta Files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "\n",
    "class FastAreader():\n",
    "    \"\"\"\n",
    "    Read in files and preprocess.\n",
    "    \"\"\"\n",
    "    def __init__(self, fname=''):\n",
    "        \"\"\" Contructor: saves attribute fname. \"\"\"\n",
    "        self.fname = fname\n",
    "\n",
    "    def doOpen(self):\n",
    "        \"\"\" Open a file.\"\"\"\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "        \"\"\" Read in a fasta file and yield header and sequences separately\"\"\"\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            # Separate headers and sequences\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 2: Missing Motif\n",
    "  \n",
    "For this assingment, I used a Markov(K-2) Model to find the probabilities of K-mers (of length K) which I assumed\n",
    "followed the binomial distribution. With these statistics I was able to print out motifs by order of their z-score (ascending order)\n",
    "and then by their length (descending order). Through the nature of this output, we can assume that the motifs \n",
    "hiding from us are those with lower z-score (closer to the top relative to motifs of the same length).\n",
    "\n",
    "### NULL Model\n",
    "- For a Markov(1) Model, we consider a preceding base\n",
    "    - Ex:  \n",
    "    \n",
    "        $P(A|T_{prev}) = \\frac{P(TA)}{P(T)} = \\frac{\\frac{c(TA)}{N}}{\\frac{c(T)}{N}} = \\frac{c(TA)}{c(T)}$  \n",
    "        \n",
    "        - N is not the same in the cases of TA and T, but the difference is negligible for now\n",
    "- For a Markov(8) Model, we consider a preceding base\n",
    "    - Ex:  \n",
    "        \n",
    "        $Pr(K) = N * E(K) = \\frac{1}{N} \\frac{c(k_{1}k_{2}k_{3}k_{4}k_{5}k_{6}k_{7}) * c(k_{2}k_{3}k_{4}k_{5}k_{6}k_{7}k_{8})}{c(k_{2}k_{3}k_{4}k_{5}k_{6}k_{7})} $  \n",
    "- Increasing the order of a Markov Model makes our model more informed\n",
    "- Check Week 2 Overview Video 3 for more details on detailed Markov Model for 4-mer described by $k_{1}k_{2}k_{3}k_{4}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "## LATEST TO WORK KINDA\n",
    "import sys\n",
    "import os \n",
    "import numpy as np\n",
    "from fastaReader import FastAreader\n",
    "\n",
    "class SearchForMissing:\n",
    "    '''\n",
    "    Initializes a length for the seqList fed as input, a dictionary, \n",
    "    a dictionary with nucleotide counts from a given list of sequences \n",
    "    (read in with FastAreader), and a list of motif (including the motif's \n",
    "    reverse complement) statistics.\n",
    "    '''\n",
    "    def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "        self.seqList = seqList\n",
    "        self.minMotif = minMotif\n",
    "        self.maxMotif = maxMotif\n",
    "        self.zCut = zCut\n",
    "        \n",
    "        self.N = self.genomeLen() \n",
    "        self.ctDict = self.ctDictMake()\n",
    "        self.output = self.outputMake()\n",
    "\n",
    "    def genomeLen(self):\n",
    "        '''\n",
    "        Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "        seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "        '''\n",
    "        totLen = 0\n",
    "        for seq in self.seqList:\n",
    "            totLen += len(seq)\n",
    "        return totLen\n",
    "\n",
    "    def ctDictMake(self):\n",
    "        '''\n",
    "        Stores a dict of dicts\n",
    "        {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "        Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "        moves over its respective sequence length and k-mers are counted.\n",
    "        '''\n",
    "        \n",
    "        finalDict = dict() # The final dict returned, stores \"collapsed counts\" of motifs and their reverse complements\n",
    "        massDict = dict() # Dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "        for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "            massDict[k] = dict()\n",
    "        for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "            finalDict[k] = dict()\n",
    "        for seq in self.seqList: # Loop through seqs in a fasta file\n",
    "            for k in range(1, self.maxMotif+1): # Loop through possible motif lengths\n",
    "                for kmer_idx in range((len(seq)-k)): # Loop through all sliding frame indices of a sequence string\n",
    "                    kmer = seq[kmer_idx:kmer_idx+k]\n",
    "                    if 'N' not in kmer: # Ignore motifs with N as a nucleotide\n",
    "                        if kmer in massDict[k]: # If motif exists in dict, increment\n",
    "                            massDict[k][kmer] += 1\n",
    "                        else: # If motif is new to dict, create an entry with value = 1\n",
    "                            massDict[k][kmer] = 1\n",
    "        # Finished composing massDict\n",
    "        # Begin composing finalDict\n",
    "        for k in range(1, self.maxMotif+1): # Loop through possible motif lengths\n",
    "            banned = set() # Create a set to hold already considered motifs, this prevents overcounting\n",
    "            for kmer in massDict[k]: # Loop through all motif entries in massDict\n",
    "                if kmer not in banned: # If motif is not considered yet, proceed\n",
    "                    banned.add(kmer)\n",
    "                    banned.add(revComp(kmer))\n",
    "                    if kmer != revComp(kmer): # If k-mer and its reverse complements are the same, collapse their counts\n",
    "                        if revComp(kmer) in massDict[k]: # If k-mer and its reverse complements exist, collapse their counts\n",
    "                            finalDict[k][kmer] = massDict[k][kmer] + massDict[k][revComp(kmer)]\n",
    "                            finalDict[k][revComp(kmer)] = massDict[k][kmer] + massDict[k][revComp(kmer)]\n",
    "                        else: # If reverse complement isn't found, but the k-mer is\n",
    "                            finalDict[k][revComp(kmer)] = massDict[k][kmer]\n",
    "                    else: # If k-mer and its reverse complements are the same, make an entry for the reverse complement with the k-mer's count, REVERSE COMPLEMENT PALINDROME\n",
    "                        finalDict[k][revComp(kmer)] = massDict[k][kmer]\n",
    "                else: # I k-mer is already considered, continue\n",
    "                    continue\n",
    "        return finalDict\n",
    "    \n",
    "    \n",
    "    def outputMake(self):\n",
    "        '''\n",
    "        Creates a list containing sorted output. This method calculates statistics such\n",
    "        as motif counts, motif expectation, motif standard deviation, and motif z-score\n",
    "        following the binomial distribution as well as a Markov(K-2) Model where K = motif length.\n",
    "        Entries are taken from ctDict keys and are transformed into output through this method.\n",
    "        '''\n",
    "        finalList = [] # List which will hold tuples of printable output \n",
    "        for k in range(self.minMotif, self.maxMotif+1):\n",
    "            banned = set()\n",
    "            for kmer in self.ctDict[k]:\n",
    "                if kmer not in banned:\n",
    "                    banned.add(kmer) \n",
    "                    banned.add(revComp(kmer))\n",
    "                    # Compute sequenceReverse entry\n",
    "                    reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "                    tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "                    sequenceReverse = (tmpList[0],tmpList[1]) # sequence: reverse entry for output\n",
    "                    # Compute count entry\n",
    "                    count = self.ctDict[k][kmer]\n",
    "                    # Compute expectation entry\n",
    "                    Lkmer = self.ctDict[k-1][kmer[:-1]] \n",
    "                    Rkmer = self.ctDict[k-1][kmer[1:]] \n",
    "                    Mkmer = self.ctDict[k-2][kmer[1:-1]]\n",
    "                    expectation = (Lkmer*Rkmer)/Mkmer # Mean following binomial distribution\n",
    "                    # Compute zscore entry\n",
    "                    numerator = count - expectation\n",
    "                    sd = (expectation * (1 - expectation/self.N))**0.5 # Standard deviation following binomial distribution\n",
    "                    zscore = numerator/sd\n",
    "                    entry = (tmpList, count, expectation, zscore, k) # k (motif length) has been included for sorting purposes\n",
    "                    # Append finished entry into list\n",
    "                    finalList.append(entry)\n",
    "                else:\n",
    "                    continue\n",
    "        finalList.sort(key=lambda x:x[-2]) # Sort by z-score\n",
    "        finalList.sort(reverse=True, key=lambda x:x[-1]) # Sort by motif length\n",
    "        return finalList\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main function \n",
    "Main function is written here. This function handles argument parsing, input, and output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "### REVERESE COMPLIMENT FUNCTION ###\n",
    "def revComp(seqStr):\n",
    "    newStr = ''\n",
    "    baseDict = {'A':'T', 'T':'A', 'C':'G', 'G':'C', 'N':'N'}\n",
    "    for char in seqStr:\n",
    "        newStr += baseDict[char]\n",
    "    return newStr[::-1]\n",
    "\n",
    "def main(infile, outfile='', inCL=None):\n",
    "    '''\n",
    "    This is the main function. Arguments are parsed here with the CommandLine class\n",
    "    provided by Dr. B. Infile and outfile are also read here, with an outfile file being \n",
    "    optional. IF no outfile is provided, output is printed. Infile is read in through the\n",
    "    readData function which returns fasta file information for the missingMotif to read.\n",
    "    This way, input only needs to be read once. missingMotif is called on a fasta file holding\n",
    "    either one or many sequences. Statistics such as observed counts, expected values, and z-score\n",
    "    are printed to output after a call to main(). Output is sorted by order of z-score (in ascending \n",
    "    order technically) and by order of motif lenght (in descending order).\n",
    "    '''\n",
    "\n",
    "    myCommandLine = CommandLine(inCL)\n",
    "    minMotif = myCommandLine.args.minMotif\n",
    "    maxMotif = myCommandLine.args.maxMotif\n",
    "    cutoff = myCommandLine.args.cutoff\n",
    "    \n",
    "    def readData(infile:'str') -> 'tuple(list,str)':\n",
    "        '''\n",
    "        Reads in fasta file as input. \n",
    "        Returns [0]: a list of seqs [1]: a string of catted seqs.\n",
    "        '''\n",
    "        seqList = []\n",
    "        readData = FastAreader(infile).readFasta()\n",
    "        for line in readData:\n",
    "            seqList.append(line[1]) # Append seq and not head to seqList\n",
    "        return seqList\n",
    "    \n",
    "    seqList = readData(infile) # Saves output of FastAreader(infile).readFasta() as seqList\n",
    "    \n",
    "    sfm = SearchForMissing(seqList, minMotif, maxMotif, cutoff)\n",
    "\n",
    "    if len(outfile) > 0:\n",
    "        with open(outfile, 'w') as myfile:\n",
    "            myfile.write(f'N = {sfm.N}\\n')\n",
    "            myfile.write(\"{0:8}:{1:8}\\t{2:5}\\t{3:5}\\t{4:1}\\n\".format(\"sequence\",\"reverse\",\"count\",\"Expect\",\"Zscore\"))\n",
    "            for line in sfm.output:\n",
    "                if line[3] < cutoff:\n",
    "                    myfile.write('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}\\n'.format(line[0][0],line[0][1], line[1], line[2],line[3]))   \n",
    "    else:        \n",
    "        print(f'N = {sfm.N}')\n",
    "        print(\"{0:8}:{1:8}\\t{2:5}\\t{3:5}\\t{4:1}\".format(\"sequence\",\"reverse\",\"count\",\"Expect\",\"Zscore\"))\n",
    "        for line in sfm.output:\n",
    "            if line[3] < cutoff:\n",
    "                print('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}'.format(line[0][0],line[0][1], line[1], line[2],line[3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    '''\n",
    "    Usage:\n",
    "        main(infile='FILE_PATH', outfile='FILE_PATH', inCL=[-ipk] )\n",
    "    Arguments:\n",
    "        --minMotif\n",
    "            Minimum motif length to consider for Markov Model\n",
    "        --maxMotif\n",
    "            Maximum motif length to consider for Markov Model\n",
    "        --cutoff\n",
    "            Significant Z-score cutoff value\n",
    "    '''\n",
    "#     main(infile='mdata/Arthrospira-platensis-NIES-39.fna',outfile='', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '-4'])\n",
    "    main(infile='mdata/Arthrospira-platensis-NIES-39.fna',outfile='Chan_Nick_MissingMotif_Output.txt', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '-4'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# INSPECTION\n",
    "\n",
    "# INSPECTION TEAM\n",
    "# Jodi Lee\n",
    "# Maxim Firsov\n",
    "# Hsiang-Yun Lu (Eloise)\n",
    "# Gabriel Aguiar\n",
    "\n",
    "# RESPONSES\n",
    "# - Markdown comments\n",
    "# - Get more docstrings\n",
    "# - Fix Null model\n",
    "# - Finish main function\n",
    "\n",
    "# CORRECTIONS\n",
    "# - Made markdown comments\n",
    "# - Made more docstrings and comments\n",
    "# - Infile is only read once in the main function\n",
    "# - Main function was put together\n",
    "# - Null model was fixed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
