{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "########################################################################\n",
    "# File: Chan_Nicholas_missingMotif.ipynb\n",
    "# Purpose: to find hidden motifs\n",
    "#   main(infile='FILE_PATH',outfile='FILE_PATH', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '0'])\n",
    "#\n",
    "# Author: Nicholas Chan\n",
    "# History: 10/18/2021 Created\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command Line Class\n",
    "Provided by Dr. B for parsing command line arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "\n",
    "\n",
    "class CommandLine():\n",
    "    \"\"\"\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse,\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination exception Usage.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, inOpts = None):\n",
    "        \"\"\"\n",
    "        CommandLine constructor.\n",
    "        \n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        \"\"\"\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "            )\n",
    "\n",
    "        self.parser.add_argument('-m', '--minMotif', type=int, default=3, action='store', help='minimum motif length to consider for Markov Model')\n",
    "        self.parser.add_argument('-M', '--maxMotif', type=int, default=8, action='store', help='maximum motif length to consider for Markov Model')\n",
    "        self.parser.add_argument('-c', '--cutoff', type=int, default=-4, action='store', help='Significant Z-score cutoff value')\n",
    "        # Command line option to use Gibbs sampling to find the optimal consensus motif.\n",
    "#         self.parser.add_argument('-g', '--gibbsampling', type=float, default=1, action='store', help='implement Gibbs sampling')\n",
    "\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FastAreader Class\n",
    "Provided by Dr. B for reading Fasta Files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "\n",
    "class FastAreader():\n",
    "    \"\"\"\n",
    "    Read in files and preprocess.\n",
    "    \"\"\"\n",
    "    def __init__(self, fname=''):\n",
    "        \"\"\" Contructor: saves attribute fname. \"\"\"\n",
    "        self.fname = fname\n",
    "\n",
    "    def doOpen(self):\n",
    "        \"\"\" Open a file.\"\"\"\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "        \"\"\" Read in a fasta file and yield header and sequences separately\"\"\"\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            # Separate headers and sequences\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 2 Notes\n",
    "\n",
    "### NULL Model\n",
    "- For a Markov(1) Model, we consider a preceding base\n",
    "    - Ex:  \n",
    "    \n",
    "        $P(A|T_{prev}) = \\frac{P(TA)}{P(T)} = \\frac{\\frac{c(TA)}{N}}{\\frac{c(T)}{N}} = \\frac{c(TA)}{c(T)}$  \n",
    "        \n",
    "        - N is not the same in the cases of TA and T, but the difference is negligible for now\n",
    "- Increasing the order of a Markov Model makes our model more informed\n",
    "- Check Week 2 Overview Video 3 for more details on detailed Markov Model for 4-mer described by $k_{1}k_{2}k_{3}k_{4}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## LATEST TO WORK KINDA\n",
    "import sys\n",
    "import os \n",
    "import numpy as np\n",
    "from fastaReader import FastAreader\n",
    "\n",
    "class SearchForMissing:\n",
    "    def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "        self.seqList = seqList\n",
    "        self.minMotif = minMotif\n",
    "        self.maxMotif = maxMotif\n",
    "        self.zCut = zCut\n",
    "        \n",
    "        self.N = self.genomeLen() # - self.maxMotif\n",
    "        self.ctDict = self.ctDictMake()\n",
    "        self.prDict = self.prDictMake()\n",
    "        self.output = self.outputMake()\n",
    "\n",
    "    def genomeLen(self):\n",
    "        '''\n",
    "        Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "        seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "        '''\n",
    "        totLen = 0\n",
    "        for seq in self.seqList:\n",
    "            totLen += len(seq)\n",
    "        return totLen\n",
    "\n",
    "    def ctDictMake(self):\n",
    "        '''\n",
    "        Stores a dict of dicts\n",
    "        {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "        Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "        moves over its respective sequence length and k-mers are counted.\n",
    "        '''\n",
    "        \n",
    "        massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "        for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "            massDict[k] = dict()\n",
    "        for seq in self.seqList:\n",
    "            for k in range(1, self.maxMotif+1):\n",
    "                for kmer_idx in range((len(seq)-k)):\n",
    "                    kmer = seq[kmer_idx:kmer_idx+k]\n",
    "                    if 'N' not in kmer:\n",
    "                        if kmer in massDict[k]:\n",
    "                            massDict[k][kmer] += 1\n",
    "                        else:\n",
    "                            massDict[k][kmer] = 1\n",
    "        return massDict\n",
    "                                \n",
    "    def prDictMake(self): # VERY DEPENDENT ON ctDict\n",
    "        massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "        for k in range(self.minMotif, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "            massDict[k] = dict()\n",
    "#         print(massDict)\n",
    "        for k in range(self.minMotif, self.maxMotif+1):\n",
    "            for kmer in self.ctDict[k].keys():\n",
    "                if (kmer[:-1] not in self.ctDict[k-1]) or (kmer[1:] not in self.ctDict[k-1]) or (kmer[1:-1] not in self.ctDict[k-2]): # If sub seq is not in sequence count \n",
    "                    massDict[k][kmer] = 0\n",
    "                else:\n",
    "                    massDict[k][kmer] = (1/self.N) * ((self.ctDict[k-1][kmer[:-1]] * self.ctDict[k-1][kmer[1:]]) / self.ctDict[k-2][kmer[1:-1]])\n",
    "        return massDict\n",
    "\n",
    "    def outputMake(self):\n",
    "        finalList = []\n",
    "        for k in range(self.minMotif, self.maxMotif+1):\n",
    "            banned = set()\n",
    "            for kmer in self.ctDict[k].keys():\n",
    "                if kmer not in banned: # interesting condition, is it implied that the reverse complement of the kmer is excluded?\n",
    "                    # sequenceReverse entry\n",
    "                    reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "                    tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "                    sequenceReverse = f\"{tmpList[0]}:{tmpList[1]}\" # sequence: reverse entry for output\n",
    "                    # count entry\n",
    "                    if kmer != reverseCompKmer: # Case 1: k-mer is not a palindrome\n",
    "                        if reverseCompKmer in self.ctDict[k]:\n",
    "                            count = self.ctDict[k][kmer] + self.ctDict[k][reverseCompKmer] \n",
    "                        else:\n",
    "                            count = self.ctDict[k][kmer]\n",
    "                    else: # Case 2: k-mer is a palindrome\n",
    "                        count = count = self.ctDict[k][kmer]\n",
    "                    # expectation entry\n",
    "                    if reverseCompKmer in self.ctDict[k]:\n",
    "#                         Pr = self.prDict[k][kmer] + self.prDict[k][reverseCompKmer] \n",
    "                        Lkmer = self.ctDict[k-1][kmer[:-1]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "                        Rkmer = self.ctDict[k-1][kmer[1:]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "                        Mkmer = self.ctDict[k-2][kmer[1:-1]]# + self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "                    else: \n",
    "                        Pr = self.prDict[k][kmer]\n",
    "                    if kmer == reverseCompKmer:    \n",
    "                        expectation =  ((Lkmer+Rkmer)**2) / (Mkmer)\n",
    "                    else:\n",
    "                        expectation =  (Lkmer*Rkmer) / (Mkmer)\n",
    "                        \n",
    "                    # zscore entry\n",
    "                    numerator = count - expectation\n",
    "                    sd = (expectation * (1 - expectation/self.N))**0.5\n",
    "                    zscore = numerator/sd\n",
    "                    entry = (sequenceReverse, count, expectation, zscore, k)\n",
    "                    banned.add(kmer) # Mark\n",
    "                    banned.add(revComp(kmer))\n",
    "                    finalList.append(entry)\n",
    "        finalList.sort(key=lambda x:x[-2]) # sort by z-score\n",
    "        finalList.sort(reverse=True, key=lambda x:x[-1]) # sort by motif length\n",
    "        return finalList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "### REVERESE COMPLIMENT FUNCTION ###\n",
    "def revComp(seqStr):\n",
    "    newStr = ''\n",
    "    baseDict = {'A':'T', 'T':'A', 'C':'G', 'G':'C', 'N':'N'}\n",
    "    for char in seqStr:\n",
    "        newStr += baseDict[char]\n",
    "    return newStr[::-1]\n",
    "\n",
    "def main(infile, outfile='', inCL=None):\n",
    "    '''\n",
    "    This is the main function. Arguments are parsed here with the CommandLine class\n",
    "    provided by Dr. B. Infile and outfile are also read here, with an outfile file being \n",
    "    optional. IF no outfile is provided, output is printed. Infile is read in through the\n",
    "    readData function which returns fasta file information for the missingMotif to read.\n",
    "    This way, input only needs to be read once. missingMotif is called on a fasta file holding\n",
    "    either one or many sequences.\n",
    "    '''\n",
    "\n",
    "    myCommandLine = CommandLine(inCL)\n",
    "    minMotif = myCommandLine.args.minMotif\n",
    "    maxMotif = myCommandLine.args.maxMotif\n",
    "    cutoff = myCommandLine.args.cutoff\n",
    "    \n",
    "    def readData(infile:'str') -> 'tuple(list,str)':\n",
    "        '''\n",
    "        Reads in fasta file as input. \n",
    "        Returns [0]: a list of seqs [1]: a string of catted seqs.\n",
    "        '''\n",
    "        seqList = []\n",
    "        readData = FastAreader(infile).readFasta()\n",
    "        for line in readData:\n",
    "            seqList.append(line[1]) # Append seq and not head to seqList\n",
    "        return seqList\n",
    "    \n",
    "    seqList = readData(infile) # Saves output of FastAreader(infile).readFasta() as seqList\n",
    "    \n",
    "    sfm = SearchForMissing(seqList, minMotif, maxMotif, cutoff)\n",
    "    \n",
    "    print(f\"N={sfm.N}\")\n",
    "    for line in sfm.output:\n",
    "        if line[3] >= cutoff:\n",
    "            break\n",
    "        print('{0:17}\\t{1:0d}\\t{2:0.2f}\\t{3:0.2f}'.format(line[0], line[1], line[2],line[3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N=6788435\n",
      "GCGATCGC:GCGATCGC\t7683\t31752.94\t-135.39\n",
      "AATTAATT:AATTAATT\t558\t2580.47\t-39.82\n",
      "AAAATTTT:AAAATTTT\t501\t2185.49\t-36.04\n",
      "TTAATTAA:TTAATTAA\t537\t2205.76\t-35.54\n",
      "AAATATTT:AAATATTT\t365\t1705.30\t-32.46\n",
      "ATAATTAT:ATAATTAT\t364\t1681.01\t-32.13\n",
      "TTTTAAAA:TTTTAAAA\t298\t1251.69\t-26.96\n",
      "CAAATTTG:CAAATTTG\t245\t1158.42\t-26.84\n",
      "TAATATTA:TAATATTA\t296\t1222.71\t-26.50\n",
      "TAAATTTA:TAAATTTA\t413\t1380.29\t-26.04\n",
      "GATTAATC:GATTAATC\t390\t1342.07\t-25.99\n",
      "TGATATCA:TGATATCA\t300\t1184.38\t-25.70\n",
      "TATTAATA:TATTAATA\t385\t1310.43\t-25.57\n",
      "ATCGCGAT:ATCGCGAT\t294\t1155.39\t-25.34\n",
      "AATATATT:AATATATT\t255\t1051.57\t-24.57\n",
      "ATTTAAAT:ATTTAAAT\t195\t945.67\t-24.41\n",
      "GTTTAAAC:GTTTAAAC\t269\t1057.29\t-24.25\n",
      "CCCTAGGG:CCCTAGGG\t232\t997.07\t-24.23\n",
      "CAATATTG:CAATATTG\t230\t971.23\t-23.79\n",
      "ACAATTGT:ACAATTGT\t115\t750.77\t-23.20\n",
      "CATTAATG:CATTAATG\t165\t800.99\t-22.47\n",
      "CTTTAAAG:CTTTAAAG\t260\t935.23\t-22.08\n",
      "TTAGCTAA:TTAGCTAA\t226\t865.22\t-21.73\n",
      "GGTTAACC:GGTTAACC\t191\t798.36\t-21.50\n",
      "TTGATCAA:TTGATCAA\t225\t851.57\t-21.47\n",
      "AGTTAACT:AGTTAACT\t174\t760.13\t-21.26\n",
      "AGATATCT:AGATATCT\t186\t776.94\t-21.20\n",
      "AAACGTTT:AAACGTTT\t89\t610.72\t-21.11\n",
      "GAATATTC:GAATATTC\t169\t740.56\t-21.00\n",
      "ATCATGAT:ATCATGAT\t201\t791.44\t-20.99\n",
      "TTTATAAA:TTTATAAA\t207\t772.49\t-20.35\n",
      "ATATATAT:ATATATAT\t253\t830.94\t-20.05\n",
      "TATATATA:TATATATA\t162\t675.77\t-19.76\n",
      "TCAATTGA:TCAATTGA\t151\t654.34\t-19.68\n",
      "TTGTACAA:TTGTACAA\t95\t550.76\t-19.42\n",
      "GAAATTTC:GAAATTTC\t233\t769.16\t-19.33\n",
      "GGATATCC:GGATATCC\t162\t652.30\t-19.20\n",
      "ATAGCTAT:ATAGCTAT\t150\t629.24\t-19.11\n",
      "TATCGATA:TATCGATA\t147\t624.39\t-19.11\n",
      "ATTATAAT:ATTATAAT\t200\t708.17\t-19.10\n",
      "TTTGCAAA:TTTGCAAA\t133\t590.67\t-18.83\n",
      "GCAATTGC:GCAATTGC\t136\t581.59\t-18.48\n",
      "CCAATTGG:CCAATTGG\t133\t575.88\t-18.46\n",
      "CTGATCAG:CTGATCAG\t117\t532.48\t-18.01\n",
      "TTATATAA:TTATATAA\t192\t650.74\t-17.98\n",
      "AATCGATT:AATCGATT\t141\t568.86\t-17.94\n",
      "CTATATAG:CTATATAG\t88\t468.63\t-17.58\n",
      "GCTATAGC:GCTATAGC\t130\t528.57\t-17.34\n",
      "TCCTAGGA:TCCTAGGA\t61\t407.10\t-17.15\n",
      "GAACGTTC:GAACGTTC\t34\t354.42\t-17.02\n",
      "GATATATC:GATATATC\t134\t517.04\t-16.85\n",
      "CTAATTAG:CTAATTAG\t98\t434.63\t-16.15\n",
      "TGTTAACA:TGTTAACA\t124\t470.63\t-15.98\n",
      "GTAATTAC:GTAATTAC\t101\t424.38\t-15.70\n",
      "ATGATCAT:ATGATCAT\t118\t451.10\t-15.68\n",
      "GGAATTCC:GGAATTCC\t127\t459.42\t-15.51\n",
      "TTCATGAA:TTCATGAA\t129\t456.50\t-15.33\n",
      "GCCATGGC:GCCATGGC\t105\t409.19\t-15.04\n",
      "CCCATGGG:CCCATGGG\t109\t409.19\t-14.84\n",
      "CGATATCG:CGATATCG\t135\t441.84\t-14.60\n",
      "TCCATGGA:TCCATGGA\t40\t271.41\t-14.05\n",
      "GTTATAAC:GTTATAAC\t72\t319.42\t-13.84\n",
      "ACTATAGT:ACTATAGT\t76\t325.81\t-13.84\n",
      "TAGCGCTA:TAGCGCTA\t116\t387.19\t-13.78\n",
      "ATTGCAAT:ATTGCAAT\t79\t324.57\t-13.63\n",
      "GTTGCAAC:GTTGCAAC\t98\t351.72\t-13.53\n",
      "CATCGATG:CATCGATG\t73\t304.48\t-13.27\n",
      "ACCTAGGT:ACCTAGGT\t69\t295.51\t-13.18\n",
      "CAACGTTG:CAACGTTG\t45\t247.87\t-12.89\n",
      "GTAGCTAC:GTAGCTAC\t59\t260.32\t-12.48\n",
      "ATGTACAT:ATGTACAT\t48\t240.75\t-12.42\n",
      "GTGATCAC:GTGATCAC\t78\t285.71\t-12.29\n",
      "ACCATGGT:ACCATGGT\t83\t293.18\t-12.28\n",
      "GATCGATC:GATCGATC\t35\t214.09\t-12.24\n",
      "CCGATCGG:CCGATCGG\t9\t164.17\t-12.11\n",
      "AGAATTCT:AGAATTCT\t51\t234.69\t-11.99\n",
      "CATATATG:CATATATG\t77\t271.26\t-11.80\n",
      "TTCGCGAA:TTCGCGAA\t66\t249.50\t-11.62\n",
      "CCTATAGG:CCTATAGG\t76\t262.36\t-11.51\n",
      "ATGGCCAT:ATGGCCAT\t48\t216.99\t-11.47\n",
      "TTGGCCAA:TTGGCCAA\t41\t204.51\t-11.43\n",
      "AAGCGCTT:AAGCGCTT\t47\t212.41\t-11.35\n",
      "CAGATCTG:CAGATCTG\t70\t248.20\t-11.31\n",
      "TCGATCGA:TCGATCGA\t10\t141.57\t-11.06\n",
      "ACGATCGT:ACGATCGT\t5\t130.26\t-10.97\n",
      "GGCGCGCC:GGCGCGCC\t66\t227.97\t-10.73\n",
      "CTGTACAG:CTGTACAG\t57\t208.79\t-10.50\n",
      "CTGGCCAG:CTGGCCAG\t36\t173.39\t-10.43\n",
      "GTCATGAC:GTCATGAC\t53\t200.87\t-10.43\n",
      "AAAGCTTT:AAAGCTTT\t53\t197.94\t-10.30\n",
      "GCGGCCGC:GCGGCCGC\t44\t183.47\t-10.30\n",
      "TAGATCTA:TAGATCTA\t48\t189.62\t-10.28\n",
      "GTCGCGAC:GTCGCGAC\t48\t184.71\t-10.06\n",
      "CCATATGG:CCATATGG\t21\t137.67\t-9.94\n",
      "TCTATAGA:TCTATAGA\t59\t196.22\t-9.80\n",
      "AGCCGGCT:AGCCGGCT\t18\t125.80\t-9.61\n",
      "ACCCGGGT:ACCCGGGT\t47\t173.00\t-9.58\n",
      "CGTTAACG:CGTTAACG\t46\t169.31\t-9.48\n",
      "CTTGCAAG:CTTGCAAG\t32\t146.01\t-9.44\n",
      "GACTAGTC:GACTAGTC\t2\t92.53\t-9.41\n",
      "CTAGCTAG:CTAGCTAG\t33\t145.12\t-9.31\n",
      "CTTATAAG:CTTATAAG\t66\t196.03\t-9.29\n",
      "TTCCGGAA:TTCCGGAA\t57\t181.79\t-9.26\n",
      "CAGCGCTG:CAGCGCTG\t27\t133.73\t-9.23\n",
      "TCACGTGA:TCACGTGA\t4\t91.86\t-9.17\n",
      "CACTAGTG:CACTAGTG\t5\t92.53\t-9.10\n",
      "AACCGGTT:AACCGGTT\t50\t167.50\t-9.08\n",
      "GGTATACC:GGTATACC\t43\t155.67\t-9.03\n",
      "TACCGGTA:TACCGGTA\t39\t149.34\t-9.03\n",
      "TGTATACA:TGTATACA\t40\t149.65\t-8.96\n",
      "TGAATTCA:TGAATTCA\t62\t182.40\t-8.91\n",
      "TCGCGCGA:TCGCGCGA\t25\t118.90\t-8.61\n",
      "CTCATGAG:CTCATGAG\t32\t130.01\t-8.60\n",
      "GTATATAC:GTATATAC\t42\t145.49\t-8.58\n",
      "AGTATACT:AGTATACT\t31\t126.75\t-8.50\n",
      "ACTGCAGT:ACTGCAGT\t12\t92.79\t-8.39\n",
      "TCCCGGGA:TCCCGGGA\t27\t118.06\t-8.38\n",
      "CCGCGCGG:CCGCGCGG\t29\t121.25\t-8.38\n",
      "AAGGCCTT:AAGGCCTT\t15\t92.12\t-8.04\n",
      "CACCGGTG:CACCGGTG\t43\t136.69\t-8.01\n",
      "GCACGTGC:GCACGTGC\t4\t71.94\t-8.01\n",
      "ACATATGT:ACATATGT\t7\t77.17\t-7.99\n",
      "ACTCGAGT:ACTCGAGT\t16\t90.37\t-7.82\n",
      "ATCCGGAT:ATCCGGAT\t47\t137.20\t-7.70\n",
      "TAACGTTA:TAACGTTA\t67\t162.44\t-7.49\n",
      "CGAATTCG:CGAATTCG\t17\t85.98\t-7.44\n",
      "AAGATCTT:AAGATCTT\t21\t91.47\t-7.37\n",
      "TAGTACTA:TAGTACTA\t41\t122.46\t-7.36\n",
      "TTCTAGAA:TTCTAGAA\t20\t88.83\t-7.30\n",
      "CAAGCTTG:CAAGCTTG\t11\t73.25\t-7.27\n",
      "GAGCGCTC:GAGCGCTC\t13\t75.06\t-7.16\n",
      "TCTCGAGA:TCTCGAGA\t12\t71.96\t-7.07\n",
      "GACGTCAA:TTGACGTC\t4\t56.63\t-6.99\n",
      "GTGGCCAC:GTGGCCAC\t18\t80.76\t-6.98\n",
      "TCGGCCGA:TCGGCCGA\t15\t73.98\t-6.86\n",
      "CTCGGGCC:GGCCCGAG\t1\t48.61\t-6.83\n",
      "GAGATCTC:GAGATCTC\t51\t127.91\t-6.80\n",
      "CGCGCGCG:CGCGCGCG\t24\t87.23\t-6.77\n",
      "TGGTACCA:TGGTACCA\t12\t67.12\t-6.73\n",
      "TGCCGGCA:TGCCGGCA\t22\t82.91\t-6.69\n",
      "TGCGCGCA:TGCGCGCA\t22\t82.77\t-6.68\n",
      "GTGTACAC:GTGTACAC\t15\t71.18\t-6.66\n",
      "GCGCGCGC:GCGCGCGC\t38\t106.43\t-6.63\n",
      "CGCCGGCG:CGCCGGCG\t15\t68.25\t-6.45\n",
      "AGCGCGCT:AGCGCGCT\t11\t60.70\t-6.38\n",
      "TGAGCTCA:TGAGCTCA\t1\t41.93\t-6.32\n",
      "CCTGCAGG:CCTGCAGG\t22\t77.59\t-6.31\n",
      "ATCTAGAT:ATCTAGAT\t28\t86.34\t-6.28\n",
      "CGGTACCG:CGGTACCG\t20\t73.86\t-6.27\n",
      "CCGGCCGG:CCGGCCGG\t11\t58.89\t-6.24\n",
      "TGGATCCA:TGGATCCA\t5\t44.63\t-5.93\n",
      "ACGGCCGT:ACGGCCGT\t28\t81.23\t-5.91\n",
      "CAGGCCTG:CAGGCCTG\t26\t77.88\t-5.88\n",
      "TAGGCCTA:TAGGCCTA\t24\t73.40\t-5.77\n",
      "CCCCGGGG:CCCCGGGG\t21\t67.47\t-5.66\n",
      "TCTGCAGA:TCTGCAGA\t11\t47.71\t-5.32\n",
      "CTCGCGAG:CTCGCGAG\t25\t68.70\t-5.27\n",
      "CTCCGGAG:CTCCGGAG\t18\t57.77\t-5.23\n",
      "GCCTAGGC:GCCTAGGC\t23\t65.22\t-5.23\n",
      "GGGGCCCC:GGGGCCCC\t10\t44.46\t-5.17\n",
      "CCTTAAGG:CCTTAAGG\t5\t35.57\t-5.13\n",
      "CGGCGCCG:CGGCGCCG\t17\t54.23\t-5.06\n",
      "ACACGTCA:TGACGTGT\t4\t32.97\t-5.05\n",
      "CGGATCCG:CGGATCCG\t11\t44.63\t-5.03\n",
      "AGGATCCT:AGGATCCT\t6\t35.27\t-4.93\n",
      "TATGCATA:TATGCATA\t20\t56.82\t-4.88\n",
      "ACTTAAGT:ACTTAAGT\t12\t44.43\t-4.87\n",
      "GGCGTCTA:TAGACGCC\t2\t26.67\t-4.78\n",
      "GCTGCAGC:GCTGCAGC\t9\t38.57\t-4.76\n",
      "GAAGCTTC:GAAGCTTC\t10\t39.34\t-4.68\n",
      "ACGCGCGT:ACGCGCGT\t3\t26.88\t-4.61\n",
      "AGGGCCCT:AGGGCCCT\t1\t23.09\t-4.60\n",
      "ATTCGAAT:ATTCGAAT\t7\t33.58\t-4.59\n",
      "GCATATGC:GCATATGC\t11\t39.37\t-4.52\n",
      "AGAGCTCT:AGAGCTCT\t8\t34.39\t-4.50\n",
      "CTCTAGAG:CTCTAGAG\t6\t31.08\t-4.50\n",
      "CCAGCTGG:CCAGCTGG\t2\t23.68\t-4.45\n",
      "CAGCTGGA:TCCAGCTG\t6\t29.34\t-4.31\n",
      "CGCATGCG:CGCATGCG\t1\t20.35\t-4.29\n",
      "AGCATGCT:AGCATGCT\t2\t22.15\t-4.28\n",
      "CACACGTC:GACGTGTG\t1\t20.17\t-4.27\n",
      "ACACGTCG:CGACGTGT\t1\t20.17\t-4.27\n",
      "ACGATCGG:CCGATCGT\t12\t37.96\t-4.21\n",
      "GTCCGGAC:GTCCGGAC\t8\t31.61\t-4.20\n",
      "GACGCCTC:GAGGCGTC\t1\t19.53\t-4.19\n",
      "GCGGCGCC:GGCGCCGC\t1\t19.53\t-4.19\n",
      "AAACATGC:GCATGTTT\t1\t19.47\t-4.19\n",
      "GAGTACTC:GAGTACTC\t20\t48.72\t-4.12\n",
      "ATTGCGCA:TGCGCAAT\t1\t17.91\t-4.00\n",
      "GCCCGGGC:GCCCGGGC\t14\t38.92\t-3.99\n",
      "GGCCGGCC:GGCCGGCC\t3\t21.00\t-3.93\n",
      "CCTCGAGG:CCTCGAGG\t7\t26.20\t-3.75\n",
      "AGGTACCT:AGGTACCT\t3\t19.52\t-3.74\n",
      "GGGATCCC:GGGATCCC\t6\t24.49\t-3.74\n",
      "TCATATGA:TCATATGA\t7\t25.83\t-3.71\n",
      "GACCGGTC:GACCGGTC\t5\t22.23\t-3.65\n",
      "GTCTAGAC:GTCTAGAC\t11\t31.08\t-3.60\n",
      "GCTTAAGC:GCTTAAGC\t15\t36.62\t-3.57\n",
      "ACGTGTAT:ATACACGT\t1\t14.56\t-3.55\n",
      "ACTAGTCC:GGACTAGT\t2\t15.82\t-3.47\n",
      "AGTGCACT:AGTGCACT\t16\t36.82\t-3.43\n",
      "ACGATCGA:TCGATCGT\t14\t33.89\t-3.42\n",
      "ACAGCTGT:ACAGCTGT\t9\t26.04\t-3.34\n",
      "CGTATACG:CGTATACG\t9\t25.69\t-3.29\n",
      "CGCTAGCG:CGCTAGCG\t13\t30.42\t-3.16\n",
      "CAGTACTG:CAGTACTG\t2\t13.40\t-3.11\n",
      "TGGGCCCA:TGGGCCCA\t5\t17.78\t-3.03\n",
      "CGGGCCCG:CGGGCCCG\t8\t22.15\t-3.01\n",
      "CCGATCGA:TCGATCGG\t21\t39.50\t-2.94\n",
      "TCTGCGCA:TGCGCAGA\t1\t10.48\t-2.93\n",
      "AACGCGTA:TACGCGTT\t1\t10.48\t-2.93\n",
      "ACATGTTC:GAACATGT\t1\t10.48\t-2.93\n",
      "AGGCGCCG:CGGCGCCT\t1\t10.48\t-2.93\n",
      "GTGCACGA:TCGTGCAC\t1\t10.48\t-2.93\n",
      "CATATGTC:GACATATG\t8\t21.59\t-2.92\n",
      "TAAGCTTA:TAAGCTTA\t12\t27.22\t-2.92\n",
      "TCTTAAGA:TCTTAAGA\t5\t16.99\t-2.91\n",
      "ACTAGTAC:GTACTAGT\t1\t10.34\t-2.90\n",
      "CTACGTGA:TCACGTAG\t1\t10.25\t-2.89\n",
      "GGGTACCC:GGGTACCC\t3\t13.65\t-2.88\n",
      "GCTCGAGC:GCTCGAGC\t6\t17.99\t-2.83\n",
      "GAGGCCTC:GAGGCCTC\t5\t16.47\t-2.83\n",
      "CGGACGTC:GACGTCCG\t1\t9.69\t-2.79\n",
      "AGACGTCC:GGACGTCT\t1\t9.69\t-2.79\n",
      "CAGACGTC:GACGTCTG\t1\t9.69\t-2.79\n",
      "GCCGCGGC:GCCGCGGC\t4\t14.69\t-2.79\n",
      "CGTGTGCC:GGCACACG\t3\t13.05\t-2.78\n",
      "GTGAAACA:TGTTTCAC\t90\t119.81\t-2.72\n",
      "TGTGCACA:TGTGCACA\t4\t14.25\t-2.72\n",
      "AGTACTCA:TGAGTACT\t3\t12.59\t-2.70\n",
      "GATGCATC:GATGCATC\t5\t15.51\t-2.67\n",
      "AGACACGC:GCGTGTCT\t26\t43.62\t-2.67\n",
      "ACTAGTAG:CTACTAGT\t4\t13.84\t-2.64\n",
      "ACGTGTGA:TCACACGT\t1\t8.87\t-2.64\n",
      "TGCTAGCA:TGCTAGCA\t2\t10.58\t-2.64\n",
      "ACGTGTAG:CTACACGT\t2\t10.48\t-2.62\n",
      "CACGTAGC:GCTACGTG\t2\t10.25\t-2.58\n",
      "ACGTACGT:ACGTACGT\t1\t8.33\t-2.54\n",
      "CACGTAGA:TCTACGTG\t4\t13.05\t-2.50\n",
      "CCGCGGGA:TCCCGCGG\t1\t8.06\t-2.49\n",
      "GGTACGTA:TACGTACC\t1\t8.06\t-2.49\n",
      "ACATGTGA:TCACATGT\t1\t8.06\t-2.49\n",
      "CATGCATG:CATGCATG\t1\t7.91\t-2.46\n",
      "ACTAGTGA:TCACTAGT\t5\t14.25\t-2.45\n",
      "AGTACTGA:TCAGTACT\t3\t11.05\t-2.42\n",
      "AATGCATT:AATGCATT\t4\t12.16\t-2.34\n",
      "CCACGTGG:CCACGTGG\t11\t21.76\t-2.31\n",
      "CACGTAGG:CCTACGTG\t3\t10.25\t-2.27\n",
      "CGACGTCG:CGACGTCG\t1\t6.67\t-2.19\n",
      "GCCCGCCC:GGGCGGGC\t80\t101.93\t-2.17\n",
      "TGACGTCA:TGACGTCA\t4\t11.27\t-2.16\n",
      "ACGTGAGG:CCTCACGT\t16\t27.28\t-2.16\n",
      "GTACGTAA:TTACGTAC\t2\t8.06\t-2.13\n",
      "AAGGCGTC:GACGCCTT\t2\t8.06\t-2.13\n",
      "AACCGCGG:CCGCGGTT\t2\t8.06\t-2.13\n",
      "ATGCGCAT:ATGCGCAT\t1\t6.37\t-2.13\n",
      "ATACGTAT:ATACGTAT\t5\t12.50\t-2.12\n",
      "CACTAGTA:TACTAGTG\t6\t13.84\t-2.11\n",
      "GGAGCTCC:GGAGCTCC\t2\t7.90\t-2.10\n",
      "ACGAGCAC:GTGCTCGT\t4\t10.92\t-2.09\n",
      "TACGTGCA:TGCACGTA\t4\t10.82\t-2.07\n",
      "CACGTGCC:GGCACGTG\t1\t6.00\t-2.04\n",
      "GGCATGTA:TACATGCC\t1\t6.00\t-2.04\n",
      "ACATGCCC:GGGCATGT\t1\t6.00\t-2.04\n",
      "GCATGTTA:TAACATGC\t2\t7.63\t-2.04\n",
      "AAGGCGCC:GGCGCCTT\t1\t5.86\t-2.01\n",
      "ATCACGTA:TACGTGAT\t5\t11.93\t-2.01\n",
      "AACGCGTT:AACGCGTT\t3\t9.00\t-2.00\n",
      "ACCGCGGT:ACCGCGGT\t2\t7.11\t-1.92\n",
      "TGCGCACA:TGTGCGCA\t1\t5.48\t-1.91\n",
      "ACGTGTCC:GGACACGT\t1\t5.48\t-1.91\n",
      "CACGTGTC:GACACGTG\t1\t5.48\t-1.91\n",
      "CCGCTAGC:GCTAGCGG\t2\t7.01\t-1.89\n",
      "TCAGCTGA:TCAGCTGA\t8\t15.34\t-1.87\n",
      "TTTCGAAA:TTTCGAAA\t3\t8.40\t-1.86\n",
      "ACTAGTCG:CGACTAGT\t5\t11.14\t-1.84\n",
      "ACGTACGC:GCGTACGT\t1\t5.12\t-1.82\n",
      "GCATGCGA:TCGCATGC\t1\t5.08\t-1.81\n",
      "GGCGCCAC:GTGGCGCC\t3\t8.06\t-1.78\n",
      "CCCGCGGG:CCCGCGGG\t10\t17.36\t-1.77\n",
      "CGGGACAC:GTGTCCCG\t36\t48.25\t-1.76\n",
      "ACACGTAC:GTACGTGT\t2\t6.47\t-1.76\n",
      "AACATGTT:AACATGTT\t3\t7.56\t-1.66\n",
      "TACTAGTA:TACTAGTA\t2\t6.09\t-1.66\n",
      "CATACGTC:GACGTATG\t16\t24.00\t-1.63\n",
      "GGCTAGCC:GGCTAGCC\t4\t8.82\t-1.62\n",
      "GGTGCACC:GGTGCACC\t4\t8.80\t-1.62\n",
      "GGACGCAC:GTGCGTCC\t104\t121.76\t-1.61\n",
      "CGCTAAGC:GCTTAGCG\t1\t4.33\t-1.60\n",
      "CGTACGCC:GGCGTACG\t1\t4.33\t-1.60\n",
      "AGGCGCCC:GGGCGCCT\t2\t5.86\t-1.59\n",
      "GAGCTCCA:TGGAGCTC\t2\t5.82\t-1.58\n",
      "CCCGCCCA:TGGGCGGG\t192\t214.99\t-1.57\n",
      "AGAGCTCG:CGAGCTCT\t1\t4.19\t-1.56\n",
      "CGACGCAC:GTGCGTCG\t70\t84.27\t-1.55\n",
      "AAGTACGT:ACGTACTT\t9\t15.00\t-1.55\n",
      "ATACATGC:GCATGTAT\t2\t5.67\t-1.54\n",
      "GCATGCCC:GGGCATGC\t1\t4.12\t-1.54\n",
      "GGCATGCA:TGCATGCC\t1\t4.12\t-1.54\n",
      "CTGCATGC:GCATGCAG\t1\t4.12\t-1.54\n",
      "AATTCGAA:TTCGAATT\t3\t7.05\t-1.53\n",
      "CGTCGACC:GGTCGACG\t1\t4.07\t-1.52\n",
      "AGGTCGAC:GTCGACCT\t1\t4.07\t-1.52\n",
      "GAGAGCTC:GAGCTCTC\t2\t5.59\t-1.52\n",
      "CGCTAGCC:GGCTAGCG\t3\t7.01\t-1.51\n",
      "CGGCTAGC:GCTAGCCG\t3\t7.01\t-1.51\n",
      "GCATGCTA:TAGCATGC\t2\t5.54\t-1.50\n",
      "GCACATGC:GCATGTGC\t2\t5.54\t-1.50\n",
      "AGCATGTG:CACATGCT\t2\t5.54\t-1.50\n",
      "AAGCATGT:ACATGCTT\t2\t5.54\t-1.50\n",
      "TGTCGACA:TGTCGACA\t1\t4.00\t-1.50\n",
      "CCTGAGGC:GCCTCAGG\t2\t5.47\t-1.48\n",
      "ATCATACG:CGTATGAT\t35\t44.93\t-1.48\n",
      "CGACGCCC:GGGCGTCG\t2\t5.45\t-1.48\n",
      "GATGCGCA:TGCGCATC\t2\t5.45\t-1.48\n",
      "AGTGCTCA:TGAGCACT\t5\t9.56\t-1.47\n",
      "CAACATGC:GCATGTTG\t1\t3.84\t-1.45\n",
      "CCGTGCCC:GGGCACGG\t19\t26.38\t-1.44\n",
      "ACACATGC:GCATGTGT\t2\t5.19\t-1.40\n",
      "CATATGGC:GCCATATG\t18\t24.90\t-1.38\n",
      "GTAGGGTA:TACCCTAC\t144\t161.52\t-1.38\n",
      "GTACGGTA:TACCGTAC\t23\t30.43\t-1.35\n",
      "AACATGCC:GGCATGTT\t1\t3.50\t-1.34\n",
      "ACATGCCG:CGGCATGT\t1\t3.50\t-1.34\n",
      "ACGTGCAT:ATGCACGT\t23\t30.34\t-1.33\n",
      "ACGTAAGT:ACTTACGT\t15\t20.66\t-1.25\n",
      "AACTAGTT:AACTAGTT\t1\t3.20\t-1.23\n",
      "GGCCCGTA:TACGGGCC\t9\t13.50\t-1.23\n",
      "GAAAGGCC:GGCCTTTC\t14\t19.33\t-1.21\n",
      "ACGTATAT:ATATACGT\t12\t16.96\t-1.20\n",
      "CACAGCTG:CAGCTGTG\t2\t4.51\t-1.18\n",
      "GGCGCCCC:GGGGCGCC\t3\t5.86\t-1.18\n",
      "CCTCTTAC:GTAAGAGG\t31\t38.16\t-1.16\n",
      "AGCTAGCT:AGCTAGCT\t3\t5.78\t-1.16\n",
      "GACGTCCC:GGGACGTC\t1\t2.98\t-1.15\n",
      "CTGCGTCA:TGACGCAG\t117\t129.85\t-1.13\n",
      "CCCGTACG:CGTACGGG\t2\t4.33\t-1.12\n",
      "GACCGGCC:GGCCGGTC\t8\t11.82\t-1.11\n",
      "CGCGTATC:GATACGCG\t16\t21.05\t-1.10\n",
      "GCTTAGCA:TGCTAAGC\t1\t2.80\t-1.08\n",
      "ACACGTAA:TTACGTGT\t1\t2.78\t-1.07\n",
      "CACGTGCG:CGCACGTG\t5\t7.92\t-1.04\n",
      "GTGCTCAA:TTGAGCAC\t22\t27.40\t-1.03\n",
      "CCGAGCAC:GTGCTCGG\t2\t4.07\t-1.02\n",
      "CCGTACAC:GTGTACGG\t15\t19.44\t-1.01\n",
      "ATACGCTA:TAGCGTAT\t39\t45.81\t-1.01\n",
      "GACATGTC:GACATGTC\t2\t4.00\t-1.00\n",
      "AGTCGACT:AGTCGACT\t2\t4.00\t-1.00\n",
      "AGCGCAGA:TCTGCGCT\t52\t59.69\t-0.99\n",
      "AAACACGT:ACGTGTTT\t4\t6.49\t-0.98\n",
      "AGCAGCTG:CAGCTGCT\t2\t3.93\t-0.97\n",
      "ATAAGAGC:GCTCTTAT\t21\t25.91\t-0.96\n",
      "CACATGCA:TGCATGTG\t3\t5.19\t-0.96\n",
      "AGGACGCC:GGCGTCCT\t1\t2.51\t-0.95\n",
      "AAGGGCGT:ACGCCCTT\t39\t45.36\t-0.94\n",
      "GCTAAGCA:TGCTTAGC\t2\t3.84\t-0.94\n",
      "CGTTCGAA:TTCGAACG\t1\t2.47\t-0.94\n",
      "ATGCATAC:GTATGCAT\t3\t5.12\t-0.94\n",
      "GGTCCGTA:TACGGACC\t14\t17.93\t-0.93\n",
      "GGGTCACC:GGTGACCC\t11\t14.48\t-0.91\n",
      "CATGCGTC:GACGCATG\t24\t28.90\t-0.91\n",
      "GCGTATGA:TCATACGC\t18\t22.25\t-0.90\n",
      "CACGTGAA:TTCACGTG\t8\t10.97\t-0.90\n",
      "CGTTCTTA:TAAGAACG\t25\t29.86\t-0.89\n",
      "GAGCACCC:GGGTGCTC\t4\t6.21\t-0.89\n",
      "CACGCATG:CATGCGTG\t8\t10.93\t-0.89\n",
      "CCCTGCCG:CGGCAGGG\t46\t52.26\t-0.87\n",
      "CGCTCTTA:TAAGAGCG\t34\t39.39\t-0.86\n",
      "AACAGCTG:CAGCTGTT\t7\t9.65\t-0.85\n",
      "GTATGTGA:TCACATAC\t27\t31.80\t-0.85\n",
      "CCCCGCCC:GGGCGGGG\t190\t201.94\t-0.84\n",
      "CTCTTAAG:CTTAAGAG\t3\t4.85\t-0.84\n",
      "CCACGTGC:GCACGTGG\t9\t11.82\t-0.82\n",
      "ACCGTGTG:CACACGGT\t22\t26.08\t-0.80\n",
      "ACGTATGC:GCATACGT\t8\t10.58\t-0.79\n",
      "AAGTGCAC:GTGCACTT\t5\t7.11\t-0.79\n",
      "CGAGCACC:GGTGCTCG\t6\t8.25\t-0.78\n",
      "CCCGCGGC:GCCGCGGG\t3\t4.67\t-0.77\n",
      "CGTACATC:GATGTACG\t23\t26.99\t-0.77\n",
      "GAGCGTGC:GCACGCTC\t21\t24.81\t-0.76\n",
      "ACATGCAG:CTGCATGT\t4\t5.82\t-0.75\n",
      "TACGTGGA:TCCACGTA\t4\t5.77\t-0.74\n",
      "ACAGCTGA:TCAGCTGT\t4\t5.75\t-0.73\n",
      "TTACGTAA:TTACGTAA\t1\t2.00\t-0.71\n",
      "CACGTGGC:GCCACGTG\t6\t7.95\t-0.69\n",
      "GAGCACGA:TCGTGCTC\t10\t12.43\t-0.69\n",
      "AGACGCAC:GTGCGTCT\t70\t75.99\t-0.69\n",
      "ACGCGTCT:AGACGCGT\t1\t1.95\t-0.68\n",
      "CACGCGTC:GACGCGTG\t1\t1.95\t-0.68\n",
      "ACTGCGCA:TGCGCAGT\t1\t1.95\t-0.68\n",
      "CGGAGCTC:GAGCTCCG\t1\t1.95\t-0.68\n",
      "ACGAGCTC:GAGCTCGT\t1\t1.95\t-0.68\n",
      "ACGTGTCA:TGACACGT\t1\t1.95\t-0.68\n",
      "ACGAACCT:AGGTTCGT\t43\t47.66\t-0.68\n",
      "GTACGTTA:TAACGTAC\t15\t17.83\t-0.67\n",
      "CAGCTAGC:GCTAGCTG\t6\t7.84\t-0.66\n",
      "CTTCGAAG:CTTCGAAG\t1\t1.88\t-0.64\n",
      "AGACGCCT:AGGCGTCT\t3\t4.31\t-0.63\n",
      "CAGCTGCC:GGCAGCTG\t3\t4.28\t-0.62\n",
      "AGCACGCT:AGCGTGCT\t23\t26.11\t-0.61\n",
      "AAGTACTC:GAGTACTT\t2\t3.07\t-0.61\n",
      "GGCACGTA:TACGTGCC\t4\t5.41\t-0.61\n",
      "CGTCGACA:TGTCGACG\t1\t1.82\t-0.61\n",
      "AAGCTTGT:ACAAGCTT\t18\t20.68\t-0.59\n",
      "CACGCGTG:CACGCGTG\t1\t1.78\t-0.58\n",
      "CGTTGCAC:GTGCAACG\t54\t58.43\t-0.58\n",
      "AAATGCAT:ATGCATTT\t5\t6.44\t-0.57\n",
      "ACGCATGC:GCATGCGT\t2\t2.98\t-0.57\n",
      "AGCTTAAG:CTTAAGCT\t6\t7.53\t-0.56\n",
      "AGTGCACA:TGTGCACT\t5\t6.40\t-0.55\n",
      "CGTCTCGA:TCGAGACG\t42\t45.71\t-0.55\n",
      "GAGCACTA:TAGTGCTC\t6\t7.49\t-0.54\n",
      "GGGTAACC:GGTTACCC\t40\t43.52\t-0.53\n",
      "CACGTTAC:GTAACGTG\t11\t12.92\t-0.53\n",
      "AATACGCT:AGCGTATT\t61\t65.27\t-0.53\n",
      "GCCGTGTA:TACACGGC\t23\t25.61\t-0.52\n",
      "CGACATGC:GCATGTCG\t2\t2.86\t-0.51\n",
      "CACGTACA:TGTACGTG\t4\t5.13\t-0.50\n",
      "CGCTTAAG:CTTAAGCG\t6\t7.35\t-0.50\n",
      "GAGCACCA:TGGTGCTC\t10\t11.68\t-0.49\n",
      "GCACGGTA:TACCGTGC\t24\t26.52\t-0.49\n",
      "ACGTGTGC:GCACACGT\t2\t2.81\t-0.48\n",
      "ACCAGCTG:CAGCTGGT\t6\t7.29\t-0.48\n",
      "TGGCGCCA:TGGCGCCA\t3\t3.90\t-0.46\n",
      "AGCTGCGC:GCGCAGCT\t33\t35.71\t-0.45\n",
      "GACGTCCA:TGGACGTC\t1\t1.57\t-0.45\n",
      "AGTCGACG:CGTCGACT\t1\t1.57\t-0.45\n",
      "AGTCGACA:TGTCGACT\t1\t1.57\t-0.45\n",
      "GCTTAAGA:TCTTAAGC\t5\t6.09\t-0.44\n",
      "GGGGTACC:GGTACCCC\t9\t10.39\t-0.43\n",
      "CCGGGCAC:GTGCCCGG\t12\t13.58\t-0.43\n",
      "CGTTCGCC:GGCGAACG\t25\t27.18\t-0.42\n",
      "ACGTACCT:AGGTACGT\t13\t14.56\t-0.41\n",
      "ACATGTAC:GTACATGT\t1\t1.50\t-0.41\n",
      "CGCATGTA:TACATGCG\t1\t1.50\t-0.41\n",
      "AGTTCGAA:TTCGAACT\t2\t2.66\t-0.40\n",
      "ACGCGTAT:ATACGCGT\t2\t2.66\t-0.40\n",
      "AGCTCCAT:ATGGAGCT\t38\t40.54\t-0.40\n",
      "CACGTGGA:TCCACGTG\t5\t5.97\t-0.40\n",
      "TATGCGCA:TGCGCATA\t1\t1.47\t-0.39\n",
      "CCGCGGCG:CGCCGCGG\t1\t1.47\t-0.39\n",
      "CGTGTGGA:TCCACACG\t19\t20.75\t-0.38\n",
      "ATGTGCAC:GTGCACAT\t3\t3.74\t-0.38\n",
      "GGTGCCTA:TAGGCACC\t8\t9.13\t-0.38\n",
      "CCTCGAGC:GCTCGAGG\t4\t4.81\t-0.37\n",
      "CAGTGCAC:GTGCACTG\t5\t5.87\t-0.36\n",
      "GCGCAGCC:GGCTGCGC\t69\t71.98\t-0.35\n",
      "ACCTGAGG:CCTCAGGT\t6\t6.92\t-0.35\n",
      "AGCTCGCG:CGCGAGCT\t13\t14.30\t-0.34\n",
      "CGGCGCCA:TGGCGCCG\t4\t4.74\t-0.34\n",
      "CGGTGCTC:GAGCACCG\t4\t4.72\t-0.33\n",
      "AAGGACAT:ATGTCCTT\t80\t83.01\t-0.33\n",
      "TATTCGAA:TTCGAATA\t4\t4.71\t-0.33\n",
      "CTGACGCC:GGCGTCAG\t5\t5.76\t-0.32\n",
      "ACGGTGTC:GACACCGT\t51\t53.28\t-0.31\n",
      "CGGAGCAC:GTGCTCCG\t2\t2.47\t-0.30\n",
      "CTGTGCTC:GAGCACAG\t8\t8.87\t-0.29\n",
      "AATACGTG:CACGTATT\t6\t6.72\t-0.28\n",
      "CCTGCCCG:CGGGCAGG\t23\t24.35\t-0.27\n",
      "CCGCGGAA:TTCCGCGG\t1\t1.30\t-0.27\n",
      "AGCTAGCG:CGCTAGCT\t1\t1.30\t-0.27\n",
      "ACGTGTGG:CCACACGT\t2\t2.41\t-0.27\n",
      "AATGCATA:TATGCATT\t6\t6.65\t-0.25\n",
      "CCTTAGGC:GCCTAAGG\t4\t4.52\t-0.24\n",
      "GCTTAGCC:GGCTAAGC\t1\t1.25\t-0.23\n",
      "ACGAGGCC:GGCCTCGT\t6\t6.58\t-0.22\n",
      "GTTACGTA:TACGTAAC\t1\t1.24\t-0.22\n",
      "CGTCGTAC:GTACGACG\t16\t16.88\t-0.21\n",
      "CCGGAACG:CGTTCCGG\t82\t83.94\t-0.21\n",
      "CTCAGGCA:TGCCTGAG\t54\t55.48\t-0.20\n",
      "CCTGCGCA:TGCGCAGG\t3\t3.35\t-0.19\n",
      "CATACGCC:GGCGTATG\t26\t26.96\t-0.18\n",
      "AGTGCCCG:CGGGCACT\t19\t19.82\t-0.18\n",
      "ACATTCGT:ACGAATGT\t40\t41.18\t-0.18\n",
      "ATGGACCG:CGGTCCAT\t14\t14.69\t-0.18\n",
      "GCCTTATA:TATAAGGC\t77\t78.50\t-0.17\n",
      "ACACCGTC:GACGGTGT\t59\t60.30\t-0.17\n",
      "CCACGTAC:GTACGTGG\t3\t3.30\t-0.17\n",
      "CGCGGGCC:GGCCCGCG\t4\t4.34\t-0.16\n",
      "GAGGACGC:GCGTCCTC\t35\t35.92\t-0.15\n",
      "AGGCGCCT:AGGCGCCT\t1\t1.16\t-0.15\n",
      "CAGACGCC:GGCGTCTG\t4\t4.31\t-0.15\n",
      "CTCCACAC:GTGTGGAG\t45\t45.94\t-0.14\n",
      "ACAGCTGC:GCAGCTGT\t4\t4.28\t-0.14\n",
      "AGGCCGGC:GCCGGCCT\t5\t5.30\t-0.13\n",
      "CTTCGTAC:GTACGAAG\t12\t12.43\t-0.12\n",
      "ACTAGTTC:GAACTAGT\t2\t2.18\t-0.12\n",
      "CGAGCGTA:TACGCTCG\t10\t10.35\t-0.11\n",
      "CACGCCTC:GAGGCGTG\t31\t31.60\t-0.11\n",
      "GCTCGACC:GGTCGAGC\t29\t29.55\t-0.10\n",
      "GATGCATA:TATGCATC\t9\t9.30\t-0.10\n",
      "ACAAGAGC:GCTCTTGT\t46\t46.68\t-0.10\n",
      "GACCCTGC:GCAGGGTC\t49\t49.69\t-0.10\n",
      "CGGGCCGA:TCGGCCCG\t14\t14.37\t-0.10\n",
      "TCGTCAGA:TCTGACGA\t92\t92.93\t-0.10\n",
      "GGACGCCA:TGGCGTCC\t4\t4.19\t-0.09\n",
      "ACATGTGG:CCACATGT\t2\t2.13\t-0.09\n",
      "CACATGTA:TACATGTG\t2\t2.13\t-0.09\n",
      "CACGTATA:TATACGTG\t5\t5.19\t-0.09\n",
      "CGTGCTCC:GGAGCACG\t5\t5.19\t-0.09\n",
      "ATGAGCTC:GAGCTCAT\t5\t5.19\t-0.08\n",
      "GCGCTAGC:GCTAGCGC\t6\t6.20\t-0.08\n",
      "AAAGTACG:CGTACTTT\t45\t45.52\t-0.08\n",
      "GTTCCGGA:TCCGGAAC\t32\t32.42\t-0.07\n",
      "ACGTTGCT:AGCAACGT\t47\t47.50\t-0.07\n",
      "GGATCCGA:TCGGATCC\t13\t13.24\t-0.07\n",
      "CCGCGGCA:TGCCGCGG\t2\t2.08\t-0.05\n",
      "ACATGCGA:TCGCATGT\t2\t2.07\t-0.05\n",
      "CCGCGGAG:CTCCGCGG\t2\t2.07\t-0.05\n",
      "GGGCGGGA:TCCCGCCC\t217\t217.65\t-0.04\n",
      "CGGCGTCC:GGACGCCG\t3\t3.07\t-0.04\n",
      "CGGGCCCA:TGGGCCCG\t5\t5.08\t-0.03\n",
      "CGAGACGA:TCGTCTCG\t58\t58.21\t-0.03\n",
      "CGCCTTGC:GCAAGGCG\t54\t54.16\t-0.02\n",
      "AGGCCACG:CGTGGCCT\t8\t8.06\t-0.02\n",
      "CCTCGTAC:GTACGAGG\t15\t15.06\t-0.02\n",
      "ACATGTCT:AGACATGT\t1\t1.01\t-0.01\n",
      "ACCGCGGA:TCCGCGGT\t1\t1.01\t-0.01\n",
      "CGCATGCA:TGCATGCG\t1\t1.01\t-0.01\n",
      "CCAGGCCG:CGGCCTGG\t10\t10.01\t-0.00\n",
      "AGCTCGGC:GCCGAGCT\t13\t13.01\t-0.00\n",
      "GGATCCAA:TTGGATCC\t13\t13.00\t-0.00\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    '''\n",
    "    This is where the main function is executed.\n",
    "    Usage:\n",
    "        main(infile='FILE_PATH', outfile='FILE_PATH', inCL=[-ipk] )\n",
    "    Arguments:\n",
    "        --minMotif\n",
    "            Minimum motif length to consider for Markov Model\n",
    "        --maxMotif\n",
    "            Maximum motif length to consider for Markov Model\n",
    "        --cutoff\n",
    "            Significant Z-score cutoff value\n",
    "    '''\n",
    "#     main(infile='mdata/Arthrospira-platensis-NIES-39.fna', inCL=['-m 3', '-M 8', '-c -4'])\n",
    "    main(infile='mdata/Arthrospira-platensis-NIES-39.fna', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# INSPECTION\n",
    "\n",
    "# INSPECTION TEAM\n",
    "# Jodi Lee\n",
    "# Maxim Firsov\n",
    "# Hsiang-Yun Lu (Eloise)\n",
    "# Gabriel Aguiar\n",
    "\n",
    "# RESPONSES\n",
    "# - Markdown comments\n",
    "# - Get more docstrings\n",
    "# - Fix Null model\n",
    "# - Finish main function\n",
    "\n",
    "# CORRECTIONS\n",
    "# - Made some markdown comments\n",
    "# - Made some more docstrings and comments\n",
    "# - Made main function print out correct format\n",
    "# - Main function was put together for the most part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
