{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "########################################################################\n",
    "# File: Chan_Nicholas_missingMotif.ipynb\n",
    "# Purpose: to find hidden motifs\n",
    "#   main(infile='FILE_PATH',outfile='FILE_PATH', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '0'])\n",
    "#\n",
    "# Author: Nicholas Chan\n",
    "# History: 10/18/2021 Created\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command Line Class\n",
    "Provided by Dr. B for parsing command line arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "\n",
    "\n",
    "class CommandLine():\n",
    "    \"\"\"\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse,\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination exception Usage.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, inOpts = None):\n",
    "        \"\"\"\n",
    "        CommandLine constructor.\n",
    "        \n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        \"\"\"\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "            )\n",
    "\n",
    "        self.parser.add_argument('-m', '--minMotif', type=int, default=3, action='store', help='minimum motif length to consider for Markov Model')\n",
    "        self.parser.add_argument('-M', '--maxMotif', type=int, default=8, action='store', help='maximum motif length to consider for Markov Model')\n",
    "        self.parser.add_argument('-c', '--cutoff', type=int, default=-4, action='store', help='Significant Z-score cutoff value')\n",
    "        # Command line option to use Gibbs sampling to find the optimal consensus motif.\n",
    "#         self.parser.add_argument('-g', '--gibbsampling', type=float, default=1, action='store', help='implement Gibbs sampling')\n",
    "\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FastAreader Class\n",
    "Provided by Dr. B for reading Fasta Files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "\n",
    "class FastAreader():\n",
    "    \"\"\"\n",
    "    Read in files and preprocess.\n",
    "    \"\"\"\n",
    "    def __init__(self, fname=''):\n",
    "        \"\"\" Contructor: saves attribute fname. \"\"\"\n",
    "        self.fname = fname\n",
    "\n",
    "    def doOpen(self):\n",
    "        \"\"\" Open a file.\"\"\"\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "        \"\"\" Read in a fasta file and yield header and sequences separately\"\"\"\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            # Separate headers and sequences\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 2 Notes\n",
    "\n",
    "### NULL Model\n",
    "- For a Markov(1) Model, we consider a preceding base\n",
    "    - Ex:  \n",
    "    \n",
    "        $P(A|T_{prev}) = \\frac{P(TA)}{P(T)} = \\frac{\\frac{c(TA)}{N}}{\\frac{c(T)}{N}} = \\frac{c(TA)}{c(T)}$  \n",
    "        \n",
    "        - N is not the same in the cases of TA and T, but the difference is negligible for now\n",
    "- Increasing the order of a Markov Model makes our model more informed\n",
    "- Check Week 2 Overview Video 3 for more details on detailed Markov Model for 4-mer described by $k_{1}k_{2}k_{3}k_{4}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "## LATEST TO WORK KINDA\n",
    "import sys\n",
    "import os \n",
    "import numpy as np\n",
    "from fastaReader import FastAreader\n",
    "\n",
    "class SearchForMissing:\n",
    "    '''\n",
    "    Initializes a length for the seqList fed as input, a dictionary...\n",
    "    '''\n",
    "    def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "        self.seqList = seqList\n",
    "        self.minMotif = minMotif\n",
    "        self.maxMotif = maxMotif\n",
    "        self.zCut = zCut\n",
    "        \n",
    "        self.N = self.genomeLen() # - self.maxMotif\n",
    "        self.ctDict = self.ctDictMake()\n",
    "        self.output = self.outputMake()\n",
    "\n",
    "    def genomeLen(self):\n",
    "        '''\n",
    "        Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "        seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "        '''\n",
    "        totLen = 0\n",
    "        for seq in self.seqList:\n",
    "            totLen += len(seq)\n",
    "        return totLen\n",
    "\n",
    "    def ctDictMake(self):\n",
    "        '''\n",
    "        Stores a dict of dicts\n",
    "        {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "        Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "        moves over its respective sequence length and k-mers are counted.\n",
    "        '''\n",
    "        \n",
    "        finalDict = dict()\n",
    "        massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "        for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "            massDict[k] = dict()\n",
    "        for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "            finalDict[k] = dict()\n",
    "        print(massDict)\n",
    "        for seq in self.seqList:\n",
    "            for k in range(1, self.maxMotif+1):\n",
    "                for kmer_idx in range((len(seq)-k)):\n",
    "                    kmer = seq[kmer_idx:kmer_idx+k]\n",
    "                    if 'N' not in kmer:\n",
    "                        if kmer in massDict[k]:\n",
    "                            massDict[k][kmer] += 1\n",
    "                        else:\n",
    "                            massDict[k][kmer] = 1\n",
    "#         print(massDict)  \n",
    "        print(massDict[8]['GCGATCGC'])\n",
    "        for k in range(1, self.maxMotif+1):\n",
    "            banned = set()\n",
    "            for kmer in massDict[k]:\n",
    "                if kmer not in banned:\n",
    "                    banned.add(kmer)\n",
    "                    banned.add(revComp(kmer))\n",
    "                    if kmer != revComp(kmer):\n",
    "                        if revComp(kmer) in massDict[k]:\n",
    "                            finalDict[k][kmer] = massDict[k][kmer] + massDict[k][revComp(kmer)]\n",
    "                            finalDict[k][revComp(kmer)] = massDict[k][kmer] + massDict[k][revComp(kmer)]\n",
    "#                         else:\n",
    "#                             continue\n",
    "#                     else:\n",
    "#                         continue\n",
    "#                 else:\n",
    "#                     continue\n",
    "#         print(finalDict)\n",
    "        return finalDict\n",
    "    \n",
    "    \n",
    "    def outputMake(self):\n",
    "        finalList = []\n",
    "        for k in range(self.minMotif, self.maxMotif+1):\n",
    "            print(f\"k in outputMake: {k}\")\n",
    "            banned = set()\n",
    "            for kmer in self.ctDict[k]:\n",
    "                print(kmer)\n",
    "                if kmer not in banned: # interesting condition, is it implied that the reverse complement of the kmer is excluded?\n",
    "                    banned.add(kmer) # Mark\n",
    "                    banned.add(revComp(kmer))\n",
    "                    # sequenceReverse entry\n",
    "                    reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "                    tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "                    sequenceReverse = f\"{tmpList[0]}:{tmpList[1]}\" # sequence: reverse entry for output\n",
    "                    # count entry\n",
    "                    count = self.ctDict[k][kmer]\n",
    "                    # expectation entry\n",
    "                    Lkmer = self.ctDict[k-1][kmer[:-1]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "                    Rkmer = self.ctDict[k-1][kmer[1:]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "                    Mkmer = self.ctDict[k-2][kmer[1:-1]]# + self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "                    if kmer == 'GCGATCGC':\n",
    "                        print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "                    expectation =  (Lkmer*Rkmer) / (Mkmer)\n",
    "                    if kmer == 'GCGATCGC':\n",
    "                        print(f\"expectation = {expectation}\")\n",
    "                        print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "                    \n",
    "                    # zscore entry\n",
    "                    numerator = count - expectation\n",
    "                    sd = (expectation * (1 - expectation/self.N))**0.5\n",
    "                    zscore = numerator/sd\n",
    "                    entry = (sequenceReverse, count, expectation, zscore, k)\n",
    "                    \n",
    "                    finalList.append(entry)\n",
    "        finalList.sort(key=lambda x:x[-2]) # sort by z-score\n",
    "        finalList.sort(reverse=True, key=lambda x:x[-1]) # sort by motif length\n",
    "        return finalList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "### REVERESE COMPLIMENT FUNCTION ###\n",
    "def revComp(seqStr):\n",
    "    newStr = ''\n",
    "    baseDict = {'A':'T', 'T':'A', 'C':'G', 'G':'C', 'N':'N'}\n",
    "    for char in seqStr:\n",
    "        newStr += baseDict[char]\n",
    "    return newStr[::-1]\n",
    "\n",
    "def main(infile, outfile='', inCL=None):\n",
    "    '''\n",
    "    This is the main function. Arguments are parsed here with the CommandLine class\n",
    "    provided by Dr. B. Infile and outfile are also read here, with an outfile file being \n",
    "    optional. IF no outfile is provided, output is printed. Infile is read in through the\n",
    "    readData function which returns fasta file information for the RandomMotifSearch to read.\n",
    "    This way, input only needs to be read once. RandomMotifSearch is called i number of times,\n",
    "    with the relative entropy score and consensus sequence being stored in the motifAndScore list.\n",
    "    The highest RE Score and its corresponding consensus sequence are written to outup.\n",
    "    '''\n",
    "\n",
    "    myCommandLine = CommandLine(inCL)\n",
    "    minMotif = myCommandLine.args.minMotif\n",
    "    maxMotif = myCommandLine.args.maxMotif\n",
    "    cutoff = myCommandLine.args.cutoff\n",
    "    \n",
    "    def readData(infile:'str') -> 'tuple(list,str)':\n",
    "        '''\n",
    "        Reads in fasta file as input. \n",
    "        Returns [0]: a list of seqs [1]: a string of catted seqs.\n",
    "        '''\n",
    "        seqList = []\n",
    "        readData = FastAreader(infile).readFasta()\n",
    "        for line in readData:\n",
    "            seqList.append(line[1]) # Append seq and not head to seqList\n",
    "        return seqList\n",
    "    \n",
    "    seqList = readData(infile) # Saves output of FastAreader(infile).readFasta() as seqList\n",
    "    \n",
    "    sfm = SearchForMissing(seqList, minMotif, maxMotif, cutoff)\n",
    "    print(sfm.ctDict[8]['GCGATCGC'])\n",
    "    print(f'N = {sfm.N}')\n",
    "    for line in sfm.output:\n",
    "        if line[3] >= cutoff:\n",
    "            break\n",
    "        print('{0:17}\\t{1:0d}\\t{2:0.2f}\\t{3:0.2f}'.format(line[0], line[1], line[2],line[3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1: {}, 2: {}, 3: {}, 4: {}, 5: {}, 6: {}, 7: {}, 8: {}}\n",
      "7683\n",
      "k in outputMake: 3\n",
      "AGC\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'GC'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-143-6ef569c22713>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     13\u001b[0m     '''\n\u001b[0;32m     14\u001b[0m \u001b[1;31m#     main(infile='mdata/Arthrospira-platensis-NIES-39.fna', inCL=['-m 3', '-M 8', '-c -4'])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 15\u001b[1;33m     \u001b[0mmain\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minfile\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'mdata/Arthrospira-platensis-NIES-39.fna'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minCL\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'--minMotif'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'3'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'--maxMotif'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'8'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'--cutoff'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'0'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-142-4c7eb318f2cc>\u001b[0m in \u001b[0;36mmain\u001b[1;34m(infile, outfile, inCL)\u001b[0m\n\u001b[0;32m     36\u001b[0m     \u001b[0mseqList\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mreadData\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minfile\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# Saves output of FastAreader(infile).readFasta() as seqList\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     37\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 38\u001b[1;33m     \u001b[0msfm\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSearchForMissing\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseqList\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mminMotif\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmaxMotif\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcutoff\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     39\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msfm\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mctDict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m8\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'GCGATCGC'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     40\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf'N = {sfm.N}'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-141-e12b2421fc08>\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, seqList, minMotif, maxMotif, zCut)\u001b[0m\n\u001b[0;32m     17\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mN\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgenomeLen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# - self.maxMotif\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     18\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mctDict\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mctDictMake\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 19\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moutput\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moutputMake\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     20\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     21\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mgenomeLen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-141-e12b2421fc08>\u001b[0m in \u001b[0;36moutputMake\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     93\u001b[0m                     \u001b[1;31m# expectation entry\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     94\u001b[0m                     \u001b[0mLkmer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mctDict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkmer\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;31m#+ self.ctDict[k-1][revComp(kmer[:-1])]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 95\u001b[1;33m                     \u001b[0mRkmer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mctDict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkmer\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;31m#+ self.ctDict[k-1][revComp(kmer[:-1])]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     96\u001b[0m                     \u001b[0mMkmer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mctDict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkmer\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;31m# + self.ctDict[k-2][revComp(kmer[1:-1])]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     97\u001b[0m                     \u001b[1;32mif\u001b[0m \u001b[0mkmer\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'GCGATCGC'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'GC'"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    '''\n",
    "    This is where the main function is executed.\n",
    "    Usage:\n",
    "        main(infile='FILE_PATH', outfile='FILE_PATH', inCL=[-ipk] )\n",
    "    Arguments:\n",
    "        --minMotif\n",
    "            Minimum motif length to consider for Markov Model\n",
    "        --maxMotif\n",
    "            Maximum motif length to consider for Markov Model\n",
    "        --cutoff\n",
    "            Significant Z-score cutoff value\n",
    "    '''\n",
    "#     main(infile='mdata/Arthrospira-platensis-NIES-39.fna', inCL=['-m 3', '-M 8', '-c -4'])\n",
    "    main(infile='mdata/Arthrospira-platensis-NIES-39.fna', inCL=['--minMotif', '3', '--maxMotif', '8', '--cutoff', '0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# INSPECTION\n",
    "\n",
    "# INSPECTION TEAM\n",
    "# Jodi Lee\n",
    "# Maxim Firsov\n",
    "# Hsiang-Yun Lu (Eloise)\n",
    "# Gabriel Aguiar\n",
    "\n",
    "# RESPONSES\n",
    "# - Markdown comments\n",
    "# - Get more docstrings\n",
    "# - Fix Null model\n",
    "# - Finish main function\n",
    "# \n",
    "\n",
    "# CORRECTIONS\n",
    "# - Made markdown comments\n",
    "# - Made more docstrings and comments\n",
    "# - Infile is only read once in the main function\n",
    "# - Main function was put together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THIS ONE SUCKS\n",
    "# import sys\n",
    "# import os \n",
    "# import numpy as np\n",
    "# from fastaReader import FastAreader\n",
    "\n",
    "# class SearchForMissing:\n",
    "#     def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "#         self.seqList = seqList\n",
    "#         self.minMotif = minMotif\n",
    "#         self.maxMotif = maxMotif\n",
    "#         self.zCut = zCut\n",
    "        \n",
    "#         self.N = self.genomeLen() # - self.maxMotif\n",
    "#         self.ctDict = self.ctDictMake()\n",
    "#         self.prDict = self.prDictMake()\n",
    "#         self.output = self.outputMake()\n",
    "\n",
    "#     def genomeLen(self):\n",
    "#         '''\n",
    "#         Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "#         seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "#         '''\n",
    "#         totLen = 0\n",
    "#         for seq in self.seqList:\n",
    "#             totLen += len(seq)\n",
    "#         return totLen\n",
    "\n",
    "#     def ctDictMake(self):\n",
    "#         '''\n",
    "#         Stores a dict of dicts\n",
    "#         {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "#         Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "#         moves over its respective sequence length and k-mers are counted.\n",
    "#         '''\n",
    "        \n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for seq in self.seqList:\n",
    "#             for k in range(1, self.maxMotif+1):\n",
    "#                 for kmer_idx in range((len(seq)-k)):\n",
    "#                     kmer = seq[kmer_idx:kmer_idx+k]\n",
    "#                     if 'N' not in kmer:\n",
    "#                         if kmer in massDict[k]:\n",
    "#                             massDict[k][kmer] += 1\n",
    "#                         else:\n",
    "#                             massDict[k][kmer] = 1\n",
    "#         return massDict\n",
    "                                \n",
    "#     def prDictMake(self): # VERY DEPENDENT ON ctDict\n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(self.minMotif, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if (kmer[:-1] not in self.ctDict[k-1]) or (kmer[1:] not in self.ctDict[k-1]) or (kmer[1:-1] not in self.ctDict[k-2]): # If sub seq is not in sequence count \n",
    "#                     massDict[k][kmer] = 0\n",
    "#                 else:\n",
    "#                     if kmer == 'AAAATTTA': # Omg realization\n",
    "#                         print(f\"KMER {kmer} count =  {self.ctDict[k][kmer]}\")\n",
    "#                         print(f\"RIGHT kmer[:-1] = {kmer[:-1]}; count = {self.ctDict[k-1][kmer[:-1]]}; revCount = {self.ctDict[k-1][revComp(kmer[:-1])]}\")\n",
    "#                         print(f\"LEFT  kmer[1:] = {kmer[1:]}; count = {self.ctDict[k-1][kmer[1:]]}; revCount = {self.ctDict[k-1][revComp(kmer[1:])]}\")\n",
    "#                         print(f\"MID kmer[1:-1] = {kmer[1:-1]}; count = {self.ctDict[k-2][kmer[1:-1]]}; revCount = {self.ctDict[k-2][revComp(kmer[1:-1])]}\")\n",
    "#                         print()\n",
    "#                     if kmer == 'TAAATTTT': # Omg realization\n",
    "#                         print(f\"KMER {kmer} count =  {self.ctDict[k][kmer]}\")\n",
    "#                         print(f\"RIGHT kmer[:-1] = {kmer[:-1]}; count = {self.ctDict[k-1][kmer[:-1]]}; revCount = {self.ctDict[k-1][revComp(kmer[:-1])]}\")\n",
    "#                         print(f\"LEFT  kmer[1:] = {kmer[1:]}; count = {self.ctDict[k-1][kmer[1:]]}; revCount = {self.ctDict[k-1][revComp(kmer[1:])]}\")\n",
    "#                         print(f\"MID kmer[1:-1] = {kmer[1:-1]}; count = {self.ctDict[k-2][kmer[1:-1]]}; revCount = {self.ctDict[k-2][revComp(kmer[1:-1])]}\")\n",
    "# #                     if kmer != revComp(kmer): #do the NEW THING HERE!!!\n",
    "#                     massDict[k][kmer] = (1/self.N) * ((self.ctDict[k-1][kmer[:-1]] * self.ctDict[k-1][kmer[1:]]) / self.ctDict[k-2][kmer[1:-1]])\n",
    "#         return massDict\n",
    "\n",
    "#     def outputMake(self):\n",
    "#         finalList = []\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             banned = set()\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if kmer not in banned: # interesting condition, is it implied that the reverse complement of the kmer is excluded?\n",
    "#                     # sequenceReverse entry\n",
    "#                     reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "#                     tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "#                     sequenceReverse = f\"{tmpList[0]}:{tmpList[1]}\" # sequence: reverse entry for output\n",
    "#                     # count entry\n",
    "#                     if kmer != reverseCompKmer: # Case 1: k-mer is not a palindrome\n",
    "#                         if reverseCompKmer in self.ctDict[k]:\n",
    "#                             count = self.ctDict[k][kmer] + self.ctDict[k][reverseCompKmer] \n",
    "#                         else:\n",
    "#                             count = self.ctDict[k][kmer]\n",
    "#                     else: # Case 2: k-mer is a palindrome\n",
    "#                         count = count = self.ctDict[k][kmer]\n",
    "#                     # expectation entry\n",
    "#                     if reverseCompKmer in self.ctDict[k]:\n",
    "# #                         Pr = self.prDict[k][kmer] + self.prDict[k][reverseCompKmer] \n",
    "#                         Lkmer = self.ctDict[k-1][kmer[:-1]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         Rkmer = self.ctDict[k-1][kmer[1:]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         Mkmer = self.ctDict[k-2][kmer[1:-1]]# + self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "#                         if kmer == 'AAAATTTA':\n",
    "#                             print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "#                         if kmer == 'TAAATTTT':\n",
    "#                             print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "# #                         if kmer == \"GCGATCGC\":\n",
    "# #                             print(f'pr of GCGATCGC: {Pr}')\n",
    "#                     else: \n",
    "#                         Pr = self.prDict[k][kmer]\n",
    "# #                     expectation = Pr * (self.N) # look at this\n",
    "# #                     Lkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Rkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Mkmer = self.ctDict[k-2][kmer[1:-1]] + (self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "#                     if kmer == reverseCompKmer:    \n",
    "#                         expectation =  ((Lkmer+Rkmer)**2) / (Mkmer)\n",
    "#                     else:\n",
    "#                         expectation =  ((Lkmer+self.ctDict[k-1][revComp(kmer[:-1])]) * (Rkmer+self.ctDict[k-1][revComp(kmer[1:])])**2) / (Mkmer)\n",
    "                        \n",
    "                        \n",
    "                        \n",
    "#                     if kmer == 'AAAATTTA':\n",
    "#                         print(f\"expectation = {expectation}\")\n",
    "#                         print(f\"(Lkmer+Rkmer)**2 = {(Lkmer+Rkmer)**2}\")\n",
    "#                         print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "#                     if kmer == 'TAAATTTT':\n",
    "#                         print(f\"expectation = {expectation}\")\n",
    "#                         print(f\"(Lkmer+Rkmer)**2 = {(Lkmer+Rkmer)**2}\")\n",
    "#                         print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")                       \n",
    "#                     # zscore entry\n",
    "#                     numerator = count - expectation\n",
    "#                     sd = (expectation * (1 - expectation/self.N))**0.5\n",
    "#                     zscore = numerator/sd\n",
    "#                     if kmer == 'AAAATTTA':\n",
    "#                         print(f\"z-score = {zscore}\")\n",
    "#                     if kmer == 'TAAATTTT':\n",
    "#                         print(f\"z-score = {zscore}\")\n",
    "#                     entry = (sequenceReverse, count, expectation, zscore, k)\n",
    "#                     banned.add(kmer) # Mark\n",
    "#                     banned.add(revComp(kmer))\n",
    "                    \n",
    "#                     finalList.append(entry)\n",
    "#         finalList.sort(key=lambda x:x[-2]) # sort by z-score\n",
    "#         finalList.sort(reverse=True, key=lambda x:x[-1]) # sort by motif length\n",
    "#         return finalList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### LATEST TO WORK KINDA\n",
    "# import sys\n",
    "# import os \n",
    "# import numpy as np\n",
    "# from fastaReader import FastAreader\n",
    "\n",
    "# class SearchForMissing:\n",
    "#     def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "#         self.seqList = seqList\n",
    "#         self.minMotif = minMotif\n",
    "#         self.maxMotif = maxMotif\n",
    "#         self.zCut = zCut\n",
    "        \n",
    "#         self.N = self.genomeLen() # - self.maxMotif\n",
    "#         self.ctDict = self.ctDictMake()\n",
    "#         self.prDict = self.prDictMake()\n",
    "#         self.output = self.outputMake()\n",
    "\n",
    "#     def genomeLen(self):\n",
    "#         '''\n",
    "#         Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "#         seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "#         '''\n",
    "#         totLen = 0\n",
    "#         for seq in self.seqList:\n",
    "#             totLen += len(seq)\n",
    "#         return totLen\n",
    "\n",
    "#     def ctDictMake(self):\n",
    "#         '''\n",
    "#         Stores a dict of dicts\n",
    "#         {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "#         Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "#         moves over its respective sequence length and k-mers are counted.\n",
    "#         '''\n",
    "        \n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for seq in self.seqList:\n",
    "#             for k in range(1, self.maxMotif+1):\n",
    "#                 for kmer_idx in range((len(seq)-k)):\n",
    "#                     kmer = seq[kmer_idx:kmer_idx+k]\n",
    "#                     if 'N' not in kmer:\n",
    "#                         if kmer in massDict[k]:\n",
    "#                             massDict[k][kmer] += 1\n",
    "#                         else:\n",
    "#                             massDict[k][kmer] = 1\n",
    "#         return massDict\n",
    "                                \n",
    "#     def prDictMake(self): # VERY DEPENDENT ON ctDict\n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(self.minMotif, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if (kmer[:-1] not in self.ctDict[k-1]) or (kmer[1:] not in self.ctDict[k-1]) or (kmer[1:-1] not in self.ctDict[k-2]): # If sub seq is not in sequence count \n",
    "#                     massDict[k][kmer] = 0\n",
    "#                 else:\n",
    "#                     if kmer == 'GCGATCGC': # Omg realization\n",
    "#                         print(f\"RIGHT kmer[:-1] = {kmer[:-1]}; count = {self.ctDict[k-1][kmer[:-1]]}\")\n",
    "#                         print(f\"LEFT  kmer[1:] = {kmer[1:]}; count = {self.ctDict[k-1][kmer[1:]]}\")\n",
    "#                         print(f\"MID kmer[1:-1] = {kmer[1:-1]}; count = {self.ctDict[k-2][kmer[1:-1]]}\")\n",
    "# #                     if kmer != revComp(kmer): #do the NEW THING HERE!!!\n",
    "#                     massDict[k][kmer] = (1/self.N) * ((self.ctDict[k-1][kmer[:-1]] * self.ctDict[k-1][kmer[1:]]) / self.ctDict[k-2][kmer[1:-1]])\n",
    "#         return massDict\n",
    "\n",
    "#     def outputMake(self):\n",
    "#         finalList = []\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             banned = set()\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if kmer not in banned: # interesting condition, is it implied that the reverse complement of the kmer is excluded?\n",
    "#                     # sequenceReverse entry\n",
    "#                     reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "#                     tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "#                     sequenceReverse = f\"{tmpList[0]}:{tmpList[1]}\" # sequence: reverse entry for output\n",
    "#                     # count entry\n",
    "#                     if kmer != reverseCompKmer: # Case 1: k-mer is not a palindrome\n",
    "#                         if reverseCompKmer in self.ctDict[k]:\n",
    "#                             count = self.ctDict[k][kmer] + self.ctDict[k][reverseCompKmer] \n",
    "#                         else:\n",
    "#                             count = self.ctDict[k][kmer]\n",
    "#                     else: # Case 2: k-mer is a palindrome\n",
    "#                         count = count = self.ctDict[k][kmer]\n",
    "#                     # expectation entry\n",
    "#                     if reverseCompKmer in self.ctDict[k]:\n",
    "# #                         Pr = self.prDict[k][kmer] + self.prDict[k][reverseCompKmer] \n",
    "#                         Lkmer = self.ctDict[k-1][kmer[:-1]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         Rkmer = self.ctDict[k-1][kmer[1:]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         Mkmer = self.ctDict[k-2][kmer[1:-1]]# + self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "#                         if kmer == 'GCGATCGC':\n",
    "#                             print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "# #                         if kmer == \"GCGATCGC\":\n",
    "# #                             print(f'pr of GCGATCGC: {Pr}')\n",
    "#                     else: \n",
    "#                         Pr = self.prDict[k][kmer]\n",
    "# #                     expectation = Pr * (self.N) # look at this\n",
    "# #                     Lkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Rkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Mkmer = self.ctDict[k-2][kmer[1:-1]] + (self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "#                     if kmer == reverseCompKmer:    \n",
    "#                         expectation =  ((Lkmer+Rkmer)**2) / (Mkmer)\n",
    "#                     else:\n",
    "#                         expectation =  (Lkmer*Rkmer) / (Mkmer)\n",
    "#                     if kmer == 'GCGATCGC':\n",
    "#                         print(f\"expectation = {expectation}\")\n",
    "#                         print(f\"(Lkmer+Rkmer)**2 = {(Lkmer+Rkmer)**2}\")\n",
    "#                         print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "                        \n",
    "#                     # zscore entry\n",
    "#                     numerator = count - expectation\n",
    "#                     sd = (expectation * (1 - expectation/self.N))**0.5\n",
    "#                     zscore = numerator/sd\n",
    "#                     entry = (sequenceReverse, count, expectation, zscore, k)\n",
    "#                     banned.add(kmer) # Mark\n",
    "#                     banned.add(revComp(kmer))\n",
    "#                     finalList.append(entry)\n",
    "#         finalList.sort(key=lambda x:x[-2]) # sort by z-score\n",
    "#         finalList.sort(reverse=True, key=lambda x:x[-1]) # sort by motif length\n",
    "#         return finalList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import sys\n",
    "# import os \n",
    "# import numpy as np\n",
    "# from fastaReader import FastAreader\n",
    "\n",
    "# class SearchForMissing:\n",
    "#     def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "#         self.seqList = seqList\n",
    "#         self.minMotif = minMotif\n",
    "#         self.maxMotif = maxMotif\n",
    "#         self.zCut = zCut\n",
    "        \n",
    "#         self.N = self.genomeLen() # - self.maxMotif\n",
    "#         self.ctDict = self.ctDictMake()\n",
    "#         self.prDict = self.prDictMake()\n",
    "#         self.output = self.outputMake()\n",
    "\n",
    "#     def genomeLen(self):\n",
    "#         '''\n",
    "#         Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "#         seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "#         '''\n",
    "#         totLen = 0\n",
    "#         for seq in self.seqList:\n",
    "#             totLen += len(seq)\n",
    "#         return totLen\n",
    "\n",
    "#     def ctDictMake(self):\n",
    "#         '''\n",
    "#         Stores a dict of dicts\n",
    "#         {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "#         Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "#         moves over its respective sequence length and k-mers are counted.\n",
    "#         '''\n",
    "        \n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for seq in self.seqList:\n",
    "#             for k in range(1, self.maxMotif+1):\n",
    "#                 for kmer_idx in range((len(seq)-k)):\n",
    "#                     kmer = seq[kmer_idx:kmer_idx+k]\n",
    "#                     if 'N' not in kmer:\n",
    "#                         if kmer in massDict[k]:\n",
    "#                             massDict[k][kmer] += 1\n",
    "#                         else:\n",
    "#                             massDict[k][kmer] = 1\n",
    "#         return massDict\n",
    "                                \n",
    "#     def prDictMake(self): # VERY DEPENDENT ON ctDict\n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(self.minMotif, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if (kmer[:-1] not in self.ctDict[k-1]) or (kmer[1:] not in self.ctDict[k-1]) or (kmer[1:-1] not in self.ctDict[k-2]): # If sub seq is not in sequence count \n",
    "#                     massDict[k][kmer] = 0\n",
    "#                 else:\n",
    "#                     if kmer == 'AAAATTTA': # Omg realization\n",
    "#                         print(f\"KMER {kmer} count =  {self.ctDict[k][kmer]}\")\n",
    "#                         print(f\"RIGHT kmer[:-1] = {kmer[:-1]}; count = {self.ctDict[k-1][kmer[:-1]]}; revCount = {self.ctDict[k-1][revComp(kmer[:-1])]}\")\n",
    "#                         print(f\"LEFT  kmer[1:] = {kmer[1:]}; count = {self.ctDict[k-1][kmer[1:]]}; revCount = {self.ctDict[k-1][revComp(kmer[1:])]}\")\n",
    "#                         print(f\"MID kmer[1:-1] = {kmer[1:-1]}; count = {self.ctDict[k-2][kmer[1:-1]]}; revCount = {self.ctDict[k-2][revComp(kmer[1:-1])]}\")\n",
    "#                         print()\n",
    "#                     if kmer == 'TAAATTTT': # Omg realization\n",
    "#                         print(f\"KMER {kmer} count =  {self.ctDict[k][kmer]}\")\n",
    "#                         print(f\"RIGHT kmer[:-1] = {kmer[:-1]}; count = {self.ctDict[k-1][kmer[:-1]]}; revCount = {self.ctDict[k-1][revComp(kmer[:-1])]}\")\n",
    "#                         print(f\"LEFT  kmer[1:] = {kmer[1:]}; count = {self.ctDict[k-1][kmer[1:]]}; revCount = {self.ctDict[k-1][revComp(kmer[1:])]}\")\n",
    "#                         print(f\"MID kmer[1:-1] = {kmer[1:-1]}; count = {self.ctDict[k-2][kmer[1:-1]]}; revCount = {self.ctDict[k-2][revComp(kmer[1:-1])]}\")\n",
    "# #                     if kmer != revComp(kmer): #do the NEW THING HERE!!!\n",
    "#                     massDict[k][kmer] = (1/self.N) * ((self.ctDict[k-1][kmer[:-1]] * self.ctDict[k-1][kmer[1:]]) / self.ctDict[k-2][kmer[1:-1]])\n",
    "#         return massDict\n",
    "\n",
    "#     def outputMake(self):\n",
    "#         finalList = []\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             banned = set()\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if kmer not in banned: # interesting condition, is it implied that the reverse complement of the kmer is excluded?\n",
    "#                     # sequenceReverse entry\n",
    "#                     reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "#                     tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "#                     sequenceReverse = f\"{tmpList[0]}:{tmpList[1]}\" # sequence: reverse entry for output\n",
    "#                     # count entry\n",
    "#                     if kmer != reverseCompKmer: # Case 1: k-mer is not a palindrome\n",
    "#                         if reverseCompKmer in self.ctDict[k]:\n",
    "#                             count = self.ctDict[k][kmer] + self.ctDict[k][reverseCompKmer] \n",
    "#                         else:\n",
    "#                             count = self.ctDict[k][kmer]\n",
    "#                     else: # Case 2: k-mer is a palindrome\n",
    "#                         count = count = self.ctDict[k][kmer]\n",
    "#                     # expectation entry\n",
    "#                     if reverseCompKmer in self.ctDict[k]:\n",
    "# #                         Pr = self.prDict[k][kmer] + self.prDict[k][reverseCompKmer] \n",
    "#                         if kmer[:-1] == revComp(kmer[:-1]):\n",
    "#                             Lkmer = self.ctDict[k-1][kmer[:-1]] \n",
    "#                         else:\n",
    "#                             Lkmer = self.ctDict[k-1][kmer[:-1]] + self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         if kmer[1:] == revComp(kmer[1:]):\n",
    "#                             Rkmer = self.ctDict[k-1][kmer[1:]]\n",
    "#                         else:\n",
    "#                             Rkmer = self.ctDict[k-1][kmer[1:]] + self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         if kmer[1:-1] == revComp(kmer[:-1]):\n",
    "#                             Mkmer = self.ctDict[k-2][kmer[1:-1]]\n",
    "#                         else:\n",
    "#                             Mkmer = self.ctDict[k-2][kmer[1:-1]] + self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "                            \n",
    "#                         if kmer == 'AAAATTTA':\n",
    "#                             print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "#                         if kmer == 'TAAATTTT':\n",
    "#                             print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "#                     else: \n",
    "#                         Pr = self.prDict[k][kmer]\n",
    "# #                     expectation = Pr * (self.N) # look at this\n",
    "# #                     Lkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Rkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Mkmer = self.ctDict[k-2][kmer[1:-1]] + (self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "# #                     if kmer == reverseCompKmer:    \n",
    "# #                         expectation =  ((Lkmer+Rkmer)**2) / (Mkmer)\n",
    "#                     if kmer == reverseCompKmer:    \n",
    "#                         expectation =  ((Lkmer+Rkmer)**2) / (Mkmer)\n",
    "#                     else:\n",
    "#                         expectation =  (Lkmer)*(Rkmer) / (Mkmer)\n",
    "                        \n",
    "                        \n",
    "                        \n",
    "#                     if kmer == 'AAAATTTA':\n",
    "#                         print(f\"expectation = {expectation}\")\n",
    "#                         print(f\"(Lkmer+Rkmer)**2 = {(Lkmer+Rkmer)**2}\")\n",
    "#                         print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "#                     if kmer == 'TAAATTTT':\n",
    "#                         print(f\"expectation = {expectation}\")\n",
    "#                         print(f\"(Lkmer+Rkmer)**2 = {(Lkmer+Rkmer)**2}\")\n",
    "#                         print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")                       \n",
    "#                     # zscore entry\n",
    "#                     numerator = count - expectation\n",
    "#                     sd = (expectation * (1 - expectation/self.N))**0.5\n",
    "#                     zscore = numerator/sd\n",
    "#                     if kmer == 'AAAATTTA':\n",
    "#                         print(f\"z-score = {zscore}\")\n",
    "#                     if kmer == 'TAAATTTT':\n",
    "#                         print(f\"z-score = {zscore}\")\n",
    "#                     entry = (sequenceReverse, count, expectation, zscore, k)\n",
    "#                     banned.add(kmer) # Mark\n",
    "#                     banned.add(revComp(kmer))\n",
    "                    \n",
    "#                     finalList.append(entry)\n",
    "#         finalList.sort(key=lambda x:x[-2]) # sort by z-score\n",
    "#         finalList.sort(reverse=True, key=lambda x:x[-1]) # sort by motif length\n",
    "#         return finalList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import sys\n",
    "# import os \n",
    "# import numpy as np\n",
    "# from fastaReader import FastAreader\n",
    "\n",
    "# class SearchForMissing:\n",
    "#     def __init__(self, seqList, minMotif, maxMotif, zCut):\n",
    "#         self.seqList = seqList\n",
    "#         self.minMotif = minMotif\n",
    "#         self.maxMotif = maxMotif\n",
    "#         self.zCut = zCut\n",
    "        \n",
    "#         self.N = self.genomeLen() # - self.maxMotif\n",
    "#         self.ctDict = self.ctDictMake()\n",
    "#         self.prDict = self.prDictMake()\n",
    "#         self.output = self.outputMake()\n",
    "\n",
    "#     def genomeLen(self):\n",
    "#         '''\n",
    "#         Takes the sum of sequence lengths from the list of sequences, seqList.\n",
    "#         seqList is made with the FastAreader class and has 1 fasta seq per entry.\n",
    "#         '''\n",
    "#         totLen = 0\n",
    "#         for seq in self.seqList:\n",
    "#             totLen += len(seq)\n",
    "#         return totLen\n",
    "\n",
    "#     def ctDictMake(self):\n",
    "#         '''\n",
    "#         Stores a dict of dicts\n",
    "#         {1:{'A':25, 'C':33,...}, ... ,8:{'AAAAGGGG':12, 'AAAATTTT':11, ...}}\n",
    "#         Reads over each sequence in the fasta. For each sequence, a sliding window\n",
    "#         moves over its respective sequence length and k-mers are counted.\n",
    "#         '''\n",
    "        \n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(1, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict; MAYBE START WITH 1 AND NOT MIN MOTIF\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for seq in self.seqList:\n",
    "#             for k in range(1, self.maxMotif+1):\n",
    "#                 for kmer_idx in range((len(seq)-k)):\n",
    "#                     kmer = seq[kmer_idx:kmer_idx+k]\n",
    "#                     if 'N' not in kmer:\n",
    "#                         if kmer in massDict[k]:\n",
    "#                             massDict[k][kmer] += 1\n",
    "#                         else:\n",
    "#                             massDict[k][kmer] = 1\n",
    "#         return massDict\n",
    "                                \n",
    "#     def prDictMake(self): # VERY DEPENDENT ON ctDict\n",
    "#         massDict = dict() # dictionary that holds dictionaries which hold k-mers and their respective counts\n",
    "#         for k in range(self.minMotif, self.maxMotif+1): # Initialize dictionaries for min to max motif sizes within massDict\n",
    "#             massDict[k] = dict()\n",
    "# #         print(massDict)\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if (kmer[:-1] not in self.ctDict[k-1]) or (kmer[1:] not in self.ctDict[k-1]) or (kmer[1:-1] not in self.ctDict[k-2]): # If sub seq is not in sequence count \n",
    "#                     massDict[k][kmer] = 0\n",
    "#                 else:\n",
    "#                     if kmer == 'GCGATCGC': # Omg realization\n",
    "#                         print(f\"RIGHT kmer[:-1] = {kmer[:-1]}; count = {self.ctDict[k-1][kmer[:-1]]}\")\n",
    "#                         print(f\"LEFT  kmer[1:] = {kmer[1:]}; count = {self.ctDict[k-1][kmer[1:]]}\")\n",
    "#                         print(f\"MID kmer[1:-1] = {kmer[1:-1]}; count = {self.ctDict[k-2][kmer[1:-1]]}\")\n",
    "# #                     if kmer != revComp(kmer): #do the NEW THING HERE!!!\n",
    "#                     massDict[k][kmer] = (1/self.N) * ((self.ctDict[k-1][kmer[:-1]] * self.ctDict[k-1][kmer[1:]]) / self.ctDict[k-2][kmer[1:-1]])\n",
    "#         return massDict\n",
    "\n",
    "#     def outputMake(self):\n",
    "#         finalList = []\n",
    "#         for k in range(self.minMotif, self.maxMotif+1):\n",
    "#             banned = set()\n",
    "#             for kmer in self.ctDict[k].keys():\n",
    "#                 if kmer not in banned: # interesting condition, is it implied that the reverse complement of the kmer is excluded?\n",
    "#                     # sequenceReverse entry\n",
    "#                     reverseCompKmer = revComp(kmer) # K-mer reverse compliment variable\n",
    "#                     tmpList = sorted([kmer,reverseCompKmer]) # Temporary list used for sorting kmer and reverse kmer for output\n",
    "#                     sequenceReverse = f\"{tmpList[0]}:{tmpList[1]}\" # sequence: reverse entry for output\n",
    "#                     # count entry\n",
    "#                     if kmer != reverseCompKmer: # Case 1: k-mer is not a palindrome\n",
    "#                         if reverseCompKmer in self.ctDict[k]:\n",
    "#                             count = self.ctDict[k][kmer] + self.ctDict[k][reverseCompKmer] \n",
    "#                         else:\n",
    "#                             count = self.ctDict[k][kmer]\n",
    "#                     else: # Case 2: k-mer is a palindrome\n",
    "#                         count = count = self.ctDict[k][kmer]\n",
    "#                     # expectation entry\n",
    "#                     if reverseCompKmer in self.ctDict[k]:\n",
    "# #                         Pr = self.prDict[k][kmer] + self.prDict[k][reverseCompKmer] \n",
    "#                         Lkmer = self.ctDict[k-1][kmer[:-1]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         Rkmer = self.ctDict[k-1][kmer[1:]] #+ self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "#                         Mkmer = self.ctDict[k-2][kmer[1:-1]]# + self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "#                         if kmer == 'GCGATCGC':\n",
    "#                             print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "# #                         if kmer == \"GCGATCGC\":\n",
    "# #                             print(f'pr of GCGATCGC: {Pr}')\n",
    "#                     else: \n",
    "#                         Pr = self.prDict[k][kmer]\n",
    "# #                     expectation = Pr * (self.N) # look at this\n",
    "# #                     Lkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Rkmer = self.ctDict[k-1][kmer[:-1]] + (self.ctDict[k-1][revComp(kmer[:-1])]\n",
    "# #                     Mkmer = self.ctDict[k-2][kmer[1:-1]] + (self.ctDict[k-2][revComp(kmer[1:-1])]\n",
    "#                     expectation =  ((Lkmer+Rkmer)**2) / (Mkmer)\n",
    "#                     if kmer == 'GCGATCGC':\n",
    "#                         print(f\"expectation = {expectation}\")\n",
    "#                         print(f\"(Lkmer+Rkmer)**2 = {(Lkmer+Rkmer)**2}\")\n",
    "#                         print(f\"Lkmer={Lkmer} Rkmer={Rkmer} Mkmer={Mkmer}\")\n",
    "                        \n",
    "#                     # zscore entry\n",
    "#                     numerator = count - expectation\n",
    "#                     sd = (expectation * (1 - expectation/self.N))**0.5\n",
    "#                     zscore = numerator/sd\n",
    "#                     entry = (sequenceReverse, count, expectation, zscore, k)\n",
    "#                     banned.add(kmer) # Mark\n",
    "#                     banned.add(revComp(kmer))\n",
    "#                     finalList.append(entry)\n",
    "#         finalList.sort(key=lambda x:x[-2]) # sort by z-score\n",
    "#         finalList.sort(reverse=True, key=lambda x:x[-1]) # sort by motif length\n",
    "#         return finalList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
