{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "########################################################################\n",
    "# File: problem16.ipynb\n",
    "\n",
    "# Author: Nicholas Chan\n",
    "# History: 11/15/2021 Created\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 5: problem 16\n",
    "<br>\n",
    "For this assignment, we were to find the nucleotide subsequences which coded <br>\n",
    "for a specified amino acid sequence. Subsequences were found from a nucleotide <br>\n",
    "sequence given on input."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Encoding Class\n",
    "<br>\n",
    "The Encoding Class is used to generate an object that can generate <br>\n",
    "all subsequences in a given sequence (input) that encode for the <br>\n",
    "amino acid sequence specified on input. Encoding class objects <br>\n",
    "look over all 6 reading frames possessed by the forward and <br>\n",
    "reverse strands of the input nucleotide sequence.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Encoding:\n",
    "    '''\n",
    "    Substring class uses seq given as input, along with storing it's \n",
    "    complement as a seqComp. The seq and seqComp are read along each \n",
    "    of their 3 reading frames, thus yielding 6 reading frames which will\n",
    "    be stored as 6 strings. \n",
    "    \n",
    "    I will use a sliding frame to determine seqs which match the amino acid seq \n",
    "    given as input. The frame will be (amino acid seq length) * 3  long and will\n",
    "    travel over every 3rd index in a seq. if a match is found, it will be stored\n",
    "    in a list.\n",
    "\n",
    "    ex input\n",
    "    ATGGCCATGGCCCCCAGAACTGAGATCAATAGTACCCGTATTAACGGGTGA\n",
    "    MA\n",
    "    '''\n",
    "    # Nucleotide to Amino Acid dictionary came from Geeks for Geeks\n",
    "    aa = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'STOP', 'TAG':'STOP',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'STOP', 'TGG':'W',\n",
    "    }\n",
    "    # Dictionary for finding nucleotide complements\n",
    "    nucComp = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C'}\n",
    "    \n",
    "    def __init__(self, nucSeq, aaSeq):\n",
    "        self.aaSeq = aaSeq\n",
    "        self.seq = nucSeq\n",
    "        self.revSeq = self.revComp(self.seq)\n",
    "        self.frLen = 3*len(self.aaSeq)\n",
    "        self.matchedSeqs = self.findAAencodings()\n",
    "\n",
    "    def revComp(self,seq):\n",
    "        ''' \n",
    "        Finds the reverse compelement of a nuc seq \n",
    "        '''\n",
    "        return ''.join([self.nucComp[nuc] for nuc in seq])[::-1]\n",
    "\n",
    "    def findSubSeqs(self, seq):\n",
    "        ''' \n",
    "        Given a seq, findSubSeqs returns a list that splits the seq\n",
    "        into seqs of length of aaSeq * 3 (num of nucs it takes to code for aaSeq)\n",
    "        '''\n",
    "        return [seq[k : self.frLen + k] for k in range(len(seq) - self.frLen + 1)]\n",
    "    \n",
    "    def findAAencodings(self):\n",
    "        '''\n",
    "        Combines all subseqs long enough to encode for aaSeq from each reading frame\n",
    "        and translates each subseq. If the subseq's amino acid translation matches with \n",
    "        aaSeq, save the seq\n",
    "        '''\n",
    "        matchedSeqs = []\n",
    "        fseqCodons = [] # List of lists containing codons of encoding length subsequences in forward and reverse strands\n",
    "        rseqCodons = [] # List of lists containing codons of encoding length subsequences in forward and reverse strands\n",
    "        for seq in self.findSubSeqs(self.seq):\n",
    "            fseqCodons.append([seq[x*3:(x+1)*3] for x in range(len(self.aaSeq))])\n",
    "        for seq in self.findSubSeqs(self.revSeq):\n",
    "            rseqCodons.append([seq[x*3:(x+1)*3] for x in range(len(self.aaSeq))])\n",
    "        for codonCombo in fseqCodons:\n",
    "            aaTrans = \"\".join([self.aa[codon] for codon in codonCombo])\n",
    "            if aaTrans == self.aaSeq:\n",
    "                matchedSeqs.append(\"\".join(codonCombo))\n",
    "        for codonCombo in rseqCodons:\n",
    "            aaTrans = \"\".join([self.aa[codon] for codon in codonCombo])\n",
    "            if aaTrans == self.aaSeq:\n",
    "                matchedSeqs.append(self.revComp(\"\".join(codonCombo)))\n",
    "        return matchedSeqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Function\n",
    "<br>\n",
    "Parses an input text file as in 2 lines. The 1st line is assumed to be a sequence of <br>\n",
    "nucleotides. The 2nd line is assumed to be a sequence of amino acids. <br>\n",
    "Input is assumed to be a string. Output is printed to stdout. <br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCCAAATGGAAACGTAACCCCTGG\n",
      "CCGAAATGGAAAAGGAACCCCTGG\n",
      "CCAAAATGGAAACGAAATCCATGG\n",
      "CCGAAATGGAAACGGAATCCCTGG\n",
      "CCCAAGTGGAAGAGAAACCCTTGG\n",
      "CCGAAATGGAAGAGAAACCCTTGG\n",
      "CCCAAGTGGAAGCGTAACCCTTGG\n",
      "CCAAAGTGGAAACGTAATCCGTGG\n",
      "CCGAAATGGAAACGGAACCCATGG\n",
      "CCAAAGTGGAAGCGCAATCCCTGG\n",
      "CCGAAGTGGAAGCGGAACCCATGG\n",
      "CCCAAATGGAAGCGGAATCCATGG\n",
      "CCAGGGATTTCGTTTCCACTTCGG\n",
      "CCATGGGTTTCTCTTCCATTTGGG\n",
      "CCAAGGATTCCGTTTCCACTTCGG\n",
      "CCACGGGTTCCTCTTCCATTTTGG\n",
      "CCAGGGATTCCGCTTCCATTTGGG\n",
      "CCACGGATTCCGTTTCCATTTCGG\n",
      "CCACGGATTCCGTTTCCATTTCGG\n",
      "CCACGGATTTCGCTTCCACTTCGG\n"
     ]
    }
   ],
   "source": [
    "def main(infile, outfile='', inCL=None):\n",
    "    '''\n",
    "    The main class reads input and genertates an Encoding object\n",
    "    called myEncoding. myEncoding is then used to generate the subsequences\n",
    "    from the nucleotide sequence given as input. These subsequences were\n",
    "    found from looking at all reading frames on the input sequence.\n",
    "    '''\n",
    "    with open(infile,'r') as myfile:\n",
    "        nucs = myfile.readline().rstrip()\n",
    "        peps = myfile.readline().rstrip()\n",
    "        myEncoding = Encoding(nucs, peps)\n",
    "        for seq in myEncoding.matchedSeqs:\n",
    "            print(seq)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(\"data/rosalind_ba4b.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# INSPECTION\n",
    "\n",
    "# INSPECTION TEAM\n",
    "# Jodi Lee\n",
    "# Nabil Mohammed\n",
    "\n",
    "# RESPONSES\n",
    "# - Use another codon to amino acid table\n",
    "# - Write more markdown comments\n",
    "# - Clean up code\n",
    "# - Add more docstrings and inline comments\n",
    "\n",
    "# CORRECTIONS\n",
    "# - Found better table to use\n",
    "# - Wrote markdown comments\n",
    "# - Cleaned code\n",
    "# - Added more docstrings and inline comments"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
