{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "########################################################################\n",
    "# File:randomizedMotifSearch.py\n",
    "# Purpose: to find the promoter motif\n",
    "#   stderr:\n",
    "#   stdout:\n",
    "#\n",
    "# Author: Nicholas Chan\n",
    "# History: 10/10/2021 Created\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "\n",
    "\n",
    "class CommandLine():\n",
    "    \"\"\"\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse,\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination exception Usage.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, inOpts = None):\n",
    "        \"\"\"\n",
    "        CommandLine constructor.\n",
    "\n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        \"\"\"\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "            )\n",
    "        # Command line option to add iteration number as input. Default: 1000.\n",
    "        self.parser.add_argument('-i', '--iterations', type=int, default=1000, action='store',\n",
    "                                 help='number of times of iterations to find consensus motif')\n",
    "\n",
    "        # Command line option to add consensus motif length as input. Default: 13.\n",
    "        self.parser.add_argument('-k', '--motifLength', type=int, default=13, action='store',\n",
    "                                 help='length of the target promoter motif')\n",
    "\n",
    "        # Command line option to add number of pseudocounts as input. Default: 1.\n",
    "        self.parser.add_argument('-p', '--pseudocount', type=float, default=1, action='store',\n",
    "                                 help='number of pseudocounts')\n",
    "\n",
    "        # Command line option to use Gibbs sampling to find the optimal consensus motif.\n",
    "        self.parser.add_argument('-g', '--gibbsampling', type=float, default=1, action='store',\n",
    "                                 help='implement Gibbs sampling')\n",
    "\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "\n",
    "class FastAreader():\n",
    "    \"\"\"\n",
    "    Read in files and preprocess.\n",
    "    \"\"\"\n",
    "    def __init__(self, fname=''):\n",
    "        \"\"\" Contructor: saves attribute fname. \"\"\"\n",
    "        self.fname = fname\n",
    "\n",
    "    def doOpen(self):\n",
    "        \"\"\" Open a file.\"\"\"\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "        \"\"\" Read in a fasta file and yield header and sequences separately\"\"\"\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            # Separate headers and sequences\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os \n",
    "import numpy as np\n",
    "from program import CommandLine\n",
    "from fastaReader import FastAreader\n",
    "\n",
    "# Handle iterations i in main function\n",
    "class RandomMotifSearch():\n",
    "    def __init__(self, k:'int', p:'', infile:'str', outfile:'str'):\n",
    "        self.infile = infile\n",
    "        self.outfile = outfile\n",
    "        self.p = p\n",
    "        self.k = k\n",
    "        self.seqList = self.readData()[0]\n",
    "        self.seqString = self.readData()[1]\n",
    "        self.nullModel = self.nullMod()\n",
    "        self.kmerPermutations = self.generateKMerPerms()\n",
    "        self.motifList = self.motifify() # First motif list\n",
    "        self.profileMatrix = self.profilify() # First motif profile matrix\n",
    "        self.REscore = self.bestProfile()\n",
    "        \n",
    "    def readData(self) -> 'tuple(list,str)':\n",
    "        '''\n",
    "        Reads in fasta file as input. \n",
    "        Returns [0]: a list of seqs [1]: a string of catted seqs.\n",
    "        '''\n",
    "        seqList = []\n",
    "        readData = FastAreader(self.infile).readFasta()\n",
    "        for line in readData:\n",
    "            seqList.append(line[1]) # Append seq and not head to seqList\n",
    "        seqString = ''.join(seqList) # After all seq lines have been added to seqList, cat all into 1 str\n",
    "        return (seqList,seqString)\n",
    "    \n",
    "    # Pseudo counts don't look like they'd make a big diff\n",
    "    def nullMod(self) -> 'dict()':\n",
    "        '''\n",
    "        Returns null model with pseudocounts.\n",
    "        '''\n",
    "        nullProfDict = dict()\n",
    "        countA = self.seqString.count('A')\n",
    "        countC = self.seqString.count('C')\n",
    "        countG = self.seqString.count('G')\n",
    "        countT = self.seqString.count('T')\n",
    "        countTot = countA + countC + countG + countT\n",
    "        nullProfDict['A'] = (countA + self.p)/(countTot + self.p*4)\n",
    "        nullProfDict['C'] = (countC + self.p)/(countTot + self.p*4)\n",
    "        nullProfDict['G'] = (countG + self.p)/(countTot + self.p*4)\n",
    "        nullProfDict['T'] = (countT + self.p)/(countTot + self.p*4)\n",
    "        return nullProfDict\n",
    "    \n",
    "    def motifify(self) -> 'list': # Randomly generates motifs, independent\n",
    "        ''' \n",
    "        Params: (self.seqList: 'list', self.k: 'int of motif length')\n",
    "        Generates n number of sequences of length k (list of k-mers/motifs). \n",
    "        In other worrds, makes a list of randomly selected \n",
    "        k-mers/motifs from given list of seqs.\n",
    "        '''\n",
    "        motifs = []\n",
    "        for seq in self.seqList:\n",
    "            indexRange = np.random.randint(0,len(seq)-self.k)\n",
    "            motifs.append(seq[indexRange:indexRange + self.k])\n",
    "        return motifs\n",
    "    \n",
    "    # Counts check out, all columns add up to 1 and fractions match decimals\n",
    "    def profilify(self) -> 'dict(lists)': # Independent\n",
    "        ''' \n",
    "        Params: (motifList: 'list', p:'int'=0)\n",
    "        Generates a profile matrix given a list of k-mers/motifs.\n",
    "        Pseudocounts are added into the Experimental and NULL models\n",
    "        in this method.\n",
    "        '''\n",
    "        # A-C-G-T correspond to 0-1-2-3\n",
    "        matRowCount = len(self.motifList)\n",
    "        motifLength = len(self.motifList[0])\n",
    "        # Dictionary matrix where rows are bases [A,C,T,G] and cols are motif indices\n",
    "        profileMatrix = dict(A=np.array([0]*motifLength), C=np.array([0]*motifLength), T=np.array([0]*motifLength), G=np.array([0]*motifLength))\n",
    "        for motif in self.motifList: # Iterate first over motifs in motif list (input)\n",
    "            idx=0 # Keep index of motif positions\n",
    "            for base in motif: # Iterate over motif bases\n",
    "                profileMatrix[base][idx]+=1\n",
    "                idx+=1\n",
    "        if self.p > 0: # If there are pseudocounts to be added\n",
    "            for key in profileMatrix.keys():\n",
    "                profileMatrix[key] = np.divide(profileMatrix[key] + self.p, matRowCount + (self.p*4)) # divides a list of motif base frequencies + pseudocount p by total bases per index + pseudocounts p*4\n",
    "            return profileMatrix\n",
    "        else: # If there are no pseudocounts to be added\n",
    "            for key in profileMatrix.keys():\n",
    "                profileMatrix[key] = np.divide(profileMatrix[key],matRowCount) # divides a list of motif base frequencies by total bases per index\n",
    "            return profileMatrix  \n",
    "        \n",
    "    def findProb(self, kmer:'str') -> 'float':\n",
    "        ''' \n",
    "        Params: (profileMatrix:'dict(lists)',kmer:'str')\n",
    "        Finds the probability of a whole k-mer given a Profile.\n",
    "        Find individual probabilities of each base in motif and finds product from each position.\n",
    "        '''\n",
    "        totProb = 1\n",
    "        for idx in range(len(kmer)):\n",
    "            totProb *= self.profileMatrix[kmer[idx]][idx]\n",
    "        return totProb\n",
    "    \n",
    "    def generateKMerPerms(self): # independent\n",
    "        ''' \n",
    "        Params: (seqList: 'List of raw sequences', k: 'int of k-mer length')\n",
    "        Generates a list of lists of all k-mer permutations from raw sequences \n",
    "        '''\n",
    "        kmerMat = []\n",
    "        for seq in self.seqList:\n",
    "            kmerList = []\n",
    "            for idx in range(len(seq)-self.k):\n",
    "                kmerList.append(seq[idx:idx+self.k])\n",
    "            kmerMat.append(kmerList)\n",
    "        return kmerMat\n",
    "    \n",
    "    def reMotifify(self) -> 'list of k-mers with the maximum probs':\n",
    "        ''''\n",
    "        Params: (kmerPermutations:'list of lists, profileMatrix:'4 x k dict of probabilities')\n",
    "        Same type of output as motifify.\n",
    "        Use this to generate motifs with maximum likelihood after initial random motif generation.\n",
    "        A PERMUTATION MATRIX IS READ IN (list of lists)\n",
    "        Each row contains all possible motif permutations per given sequence.\n",
    "        '''\n",
    "        maxPermList = []\n",
    "        for permList in self.kmerPermutations:\n",
    "            probList = []\n",
    "            for perm in permList:\n",
    "                probList.append(self.findProb(perm)) # Calculates probability of a motif given a profile and appends the probability\n",
    "            maxPermList.append(permList[probList.index(max(probList))]) # Finds index of the motif with the highest probability in a PERMUTATION LIST and appends the permutation to maxPermList\n",
    "        return maxPermList\n",
    "\n",
    "    def findConsensus(self): # Can be called at end AFTER calling rms.bestProfile()\n",
    "        '''\n",
    "        Params: (profMat:'dict of np arrays')\n",
    "        Finds a consensus sequence from the current profile matrix.\n",
    "        '''\n",
    "        motifLength = len(self.profileMatrix['A'])\n",
    "        baseList = ['A','C','T','G']\n",
    "        conSeq = ''\n",
    "        for idx in range(motifLength):\n",
    "            baseColList = []\n",
    "            for base in baseList:\n",
    "                baseColList.append((base,self.profileMatrix[base][idx]))\n",
    "            # print(baseColList,end=' ')\n",
    "            baseColList.sort(key=lambda x:x[1])\n",
    "            # print(baseColList,end='\\n\\n')\n",
    "            conSeq += baseColList[::-1][0][0] # Way of getting to value as baseColList looks like [['A',1],['B',2], ...,['F',6]]\n",
    "        return conSeq\n",
    "    \n",
    "    def realEntropy(self, expModel:'dict(lists)'):\n",
    "        '''\n",
    "        Exp model is the profile matrix you want to score\n",
    "        '''\n",
    "        colLen = len(expModel['A'])\n",
    "        bases = ['A', 'T', 'C', 'G']\n",
    "        fullPart = []# list for sum of cols of RE eq\n",
    "        for col in range(colLen): # calculation for RE for 1 column\n",
    "            basePart = [] # list for inner sum of RE eq\n",
    "            for base in bases: # calculation for RE for 1 base\n",
    "                frac = expModel[base][col]/self.nullModel[base]\n",
    "                #print(frac)\n",
    "                logged = np.log2(frac)\n",
    "                bpart = expModel[base][col]*logged\n",
    "                basePart.append(bpart)\n",
    "            fullPart.append(sum(basePart))\n",
    "            #print(sum(basePart))\n",
    "        return sum(fullPart)       \n",
    "    \n",
    "    def bestProfile(self):\n",
    "        RE = self.realEntropy(self.profileMatrix)\n",
    "#         print(\"RE 1: \", end = ' ')\n",
    "#         print(RE)\n",
    "        diff = 0\n",
    "        First = True\n",
    "        while (diff > 0.0 or First):\n",
    "            self.motifList = self.reMotifify()\n",
    "            self.profileMatrix = self.profilify()\n",
    "#             print(self.realEntropy(self.profileMatrix))\n",
    "            First = False\n",
    "            diff = self.realEntropy(self.profileMatrix) - RE\n",
    "            RE = self.realEntropy(self.profileMatrix)\n",
    "#             print(\"RE 2: \", end = ' ')\n",
    "#             print(RE)\n",
    "        return RE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shannon's Entropy\n",
    "$H(x) = -\\sum_{i = 1}^{n} P(x_{i}) log_{2} (P(x_{i}))$ \n",
    "<br/>\n",
    "$H(x) = -\\sum_{i \\in [A,C,G,T]} P(x_{i}) log_{2} (P(x_{i}))$\n",
    "- An encoding cost\n",
    "\n",
    "## Relative Entropy\n",
    "$D_{KL}(P||Q) = \\sum_{i \\in [A,C,G,T]} P(x_{i}) log_{2} \\frac{P(x_{i})} {Q(x_{i})}$\n",
    "- A distance between 2 profile probability distributions\n",
    "- P and Q are 2 probability profile distributions\n",
    "- We look at the log of a ratio of 2 probabilities (a comparison)\n",
    "- We want to compare our observed profile to just the base composition of all our sequences\n",
    "- Model Q is our NULL model\n",
    "    - This is the probability of each of our 4 bases\n",
    "- Model P describes our experimental model\n",
    "    - This is the probability of each of our 4 bases in our selected motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': array([0.66666667, 0.55555556, 0.55555556, 0.55555556, 0.44444444,\n",
      "       0.11111111, 0.22222222, 0.66666667, 0.66666667, 0.66666667,\n",
      "       0.66666667, 0.11111111, 0.55555556]), 'C': array([0.11111111, 0.22222222, 0.22222222, 0.22222222, 0.33333333,\n",
      "       0.66666667, 0.55555556, 0.11111111, 0.11111111, 0.11111111,\n",
      "       0.11111111, 0.33333333, 0.22222222]), 'T': array([0.11111111, 0.11111111, 0.11111111, 0.11111111, 0.11111111,\n",
      "       0.11111111, 0.11111111, 0.11111111, 0.11111111, 0.11111111,\n",
      "       0.11111111, 0.11111111, 0.11111111]), 'G': array([0.11111111, 0.11111111, 0.11111111, 0.11111111, 0.11111111,\n",
      "       0.11111111, 0.11111111, 0.11111111, 0.11111111, 0.11111111,\n",
      "       0.11111111, 0.44444444, 0.11111111])}\n"
     ]
    }
   ],
   "source": [
    "aa =  RandomMotifSearch(k=13, p=1, infile=\"mdata/pogCrisprs\",outfile='placeholder')\n",
    "huh = [aa.seqList, aa.seqString, aa.nullModel, aa.motifList, aa.profileMatrix, aa.kmerPermutations]\n",
    "print(huh[4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "li = []\n",
    "for i in range(1000):\n",
    "    bb =  RandomMotifSearch(k=13, p=1, infile=\"mdata/p1860Crisprs\",outfile='carbara')\n",
    "    bb.bestProfile()\n",
    "    #print(bb.bestProfile(), bb.findConsensus())\n",
    "    li.append([bb.bestProfile(), bb.findConsensus()])\n",
    "li.sort(key=lambda x:x[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[8.217968381511609, 'GAAAAACTTAAAA']"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "li[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[8.657886340591505, 'CGAAAAACTTAAA']"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "li[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
